---
title: "Visualizing Differential Expression Results"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output: html_document
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
set.seed("1492")
```

# Getting Things Setup

As usual, make sure we have the right packages for this exercise

```{r ready-packages, results="hide",message=FALSE}
if (!require("pacman")) install.packages("pacman"); library(pacman)

# let's load all of the files we were using and want to have again today
p_load("tidyverse", "knitr", "readr",
       "pander", "BiocManager", 
       "dplyr", "stringr", 
       "purrr", # for working with lists (beautify column names)
       "scales", "viridis", # for ggplot
       "reactable") # for pretty tables.

# We also need these packages today.
p_load("DESeq2", "edgeR", "AnnotationDbi", "org.Sc.sgd.db",
       "ggrepel",
       "Glimma",
       "ggVennDiagram", "ggplot2")
```

# Description

This exercises shows more ways differential expression analysis data can be visualized.

# Learning outcomes

At the end of this exercise, you should be able to:

-   Visualize Differential Expression Results


```{r load-libraries}
library(org.Sc.sgd.db)
```

```{r compare-DEworkflows}
# load in all of the DE results for the difference of difference contrast
path_output_edgeR <- "~/Desktop/Genomic_Data_Analysis/Analysis/edgeR/yeast_topTags_edgeR.tsv"
path_output_DESeq2 <- "~/Desktop/Genomic_Data_Analysis/Analysis/DESeq2/yeast_res_DESeq2.tsv"
path_output_limma <- "~/Desktop/Genomic_Data_Analysis/Analysis/limma/yeast_topTags_limma.tsv"

topTags_edgeR <- read_tsv(path_output_edgeR)
topTags_DESeq2 <- read_tsv(path_output_DESeq2)
topTags_limma <- read_tsv(path_output_limma)
```

```{r}
sig_cutoff <- 0.05
FC_cutoff <- 1

## edgeR
# get genes that are upregualted
up_edgeR_DEG <- topTags_edgeR %>%
  dplyr::filter(FDR < sig_cutoff & logFC > FC_cutoff) %>%
  pull(ORF)

down_edgeR_DEG <- topTags_edgeR %>%
  dplyr::filter(FDR < sig_cutoff & logFC < -FC_cutoff) %>%
  pull(ORF)

## DESeq2
up_DESeq2_DEG <- topTags_DESeq2 %>%
  dplyr::filter(padj < sig_cutoff & log2FoldChange > FC_cutoff) %>%
  pull(ORF)

down_DESeq2_DEG <- topTags_DESeq2 %>%
  dplyr::filter(padj < sig_cutoff & log2FoldChange < -FC_cutoff) %>%
  pull(ORF)

## limma
up_limma_DEG <- topTags_limma %>%
  dplyr::filter(adj.P.Val < sig_cutoff & logFC > FC_cutoff) %>%
  pull(ORF)

down_limma_DEG <- topTags_limma %>%
  dplyr::filter(adj.P.Val < sig_cutoff & logFC < -FC_cutoff) %>%
  pull(ORF)

up_DEG_results_list <- list(up_edgeR_DEG,
                        up_DESeq2_DEG,
                        up_limma_DEG)
```

# MA-plot

MA plots display a log ratio (M) vs an average (A) in order to visualize the differences between two groups. In general we would expect the expression of genes to remain consistent between conditions and so the MA plot should be similar to the shape of a trumpet with most points residing on a y intercept of 0. DESeq2 has a built in function for creating the MA plot that we have used before (`plotMA()`), but we can also make our own:

```{r plot-MA, out.width="33.3%"}
topTags_edgeR %>%
  mutate(`Significant FDR` = case_when(
        FDR < sig_cutoff ~ "Yes",
        .default = "No"),
        delabel = case_when(FDR < sig_cutoff & abs(logFC) > FC_label_cutoff ~ ORF,
                             .default = NA)) %>%
  ggplot(aes(x=logCPM, y=logFC, color = `Significant FDR`, label = delabel)) + 
    geom_point(size=1) + 
    scale_y_continuous(limits=c(-5, 5), oob=squish) + 
    geom_hline(yintercept = 0, colour="darkgrey", linewidth=1, linetype="longdash") +
    labs(x="mean of normalized counts", y="log fold change") +
    # ggrepel::geom_text_repel(size = 1.5) +
    scale_color_manual(values = c("black", "red")) +
    theme_bw() + 
    ggtitle("edgeR MA plot")


topTags_DESeq2 %>%
  mutate(
    `Significant FDR` = case_when(padj < sig_cutoff ~ "Yes",
                                  .default = "No"),
    delabel = case_when(
      padj < sig_cutoff & abs(log2FoldChange) > FC_label_cutoff ~ ORF,
      .default = NA)
  ) %>% 
  ggplot(aes(log(baseMean), log2FoldChange, color = `Significant FDR`, label = delabel)) +
    geom_point(size=1) + 
    scale_y_continuous(limits=c(-5, 5), oob=squish) + 
    geom_hline(yintercept = 0, colour="darkgrey", linewidth=1, linetype="longdash") +
    labs(x="mean of normalized counts", y="log fold change") +
    # ggrepel::geom_text_repel(size = 1.5) +
    scale_color_manual(values = c("black", "red")) +
    theme_bw() +
    ggtitle("DESeq2 MA plot")
  

topTags_limma %>%
  mutate(
    `Significant FDR` = case_when(adj.P.Val < sig_cutoff ~ "Yes",
                                  .default = "No"),
    delabel = case_when(
      adj.P.Val < sig_cutoff & abs(logFC) > FC_label_cutoff ~ ORF,
      .default = NA)
  ) %>% 
  ggplot(aes(AveExpr, logFC, color = `Significant FDR`, label = delabel)) + 
    geom_point(size=1) + 
    scale_y_continuous(limits=c(-5, 5), oob=squish) + 
    geom_hline(yintercept = 0, colour="darkgrey", linewidth=1, linetype="longdash") +
    labs(x="mean of normalized counts", y="log fold change") + 
    # ggrepel::geom_text_repel(size = 1.5) +
    scale_color_manual(values = c("black", "red")) +
    theme_bw() + 
    ggtitle("limma MA plot")
```


```{r plot-volcano, out.width="33.3%"}
# change the dimensions of the output figure by clicking the gear icon in topright corner of the code chunk > "use custom figure size"

# assign pvalue and logFC cutoffs for coloring DE genes
sig_cutoff <- 0.01
FC_label_cutoff <- 3


topTags_edgeR %>%
  mutate(`Significant FDR` = case_when(
        FDR < sig_cutoff ~ "Yes",
        .default = "No"),
        delabel = case_when(FDR < sig_cutoff & abs(logFC) > FC_label_cutoff ~ ORF,
                             .default = NA)) %>%
  ggplot(aes(x = logFC, -log10(FDR), color = `Significant FDR`, label = delabel)) +
  geom_point(size = 1) +
  ggrepel::geom_text_repel(size = 1.5) +
  labs(x = "log fold change", y = "-log10(adjusted p-value)") +
  theme_bw() +
  guides(color="none") +
  scale_color_manual(values = c("black", "red")) +
  ggtitle("edgeR Volcano plot")


topTags_DESeq2 %>%
  mutate(
    `Significant FDR` = case_when(padj < sig_cutoff ~ "Yes",
                                  .default = "No"),
    delabel = case_when(
      padj < sig_cutoff & abs(log2FoldChange) > FC_label_cutoff ~ ORF,
      .default = NA)
  ) %>% 
  ggplot(aes(log2FoldChange,-log10(padj), color = `Significant FDR`, label = delabel)) + 
    geom_point(size = 1) +
  ggrepel::geom_text_repel(size = 1.5) +
  labs(x = "log fold change", y = "-log10(adjusted p-value)") +
  theme_bw() +
  guides(color="none") +
  scale_color_manual(values = c("black", "red")) +
    ggtitle("DESeq2 Volcano plot")


topTags_limma %>%
  mutate(
    `Significant FDR` = case_when(adj.P.Val < sig_cutoff ~ "Yes",
                                  .default = "No"),
    delabel = case_when(
      adj.P.Val < sig_cutoff & abs(logFC) > FC_label_cutoff ~ ORF,
      .default = NA)
  ) %>% 
  ggplot(aes(x=logFC, y=-log10(P.Value), color = `Significant FDR`, label = delabel)) + 
  geom_point(size = 1) +
  ggrepel::geom_text_repel(size = 1.5) +
  labs(x = "log fold change", y = "-log10(UNADJUSTED p-value)") +
  theme_bw() +
  guides(color="none") +
  scale_color_manual(values = c("black", "red")) +
  ggtitle("limma Volcano plot")
```


# Using Glimma for an interactive visualization

## MA plots

```{r}
# load in res objects for both limma and edgeR
res_limma <- readRDS("~/Desktop/Genomic_Data_Analysis/Analysis/limma/yeast_res_limma.Rds")
res_edgeR <- readRDS("~/Desktop/Genomic_Data_Analysis/Analysis/edgeR/yeast_res_edgeR.Rds")

# load in the DGE lists for each
y_limma <- readRDS("~/Desktop/Genomic_Data_Analysis/Analysis/edgeR/yeast_res_limma.Rds")

y_edgeR <- readRDS("~/Desktop/Genomic_Data_Analysis/Analysis/edgeR/yeast_res_edgeR.Rds")
  

glimmaMA(res_limma, dge = y_limma)
glimmaMA(res_edgeR, dge = y_edgeR)

```


## Volcano Plots


```{r}
glimmaVolcano(res_limma)
glimmaVolcano(res_edgeR)
```


# Generating bargraph summaries

This really compresses relevant information, so it's generally not a good way to visualize DE data. However, it is done, so if it is useful for your experiment, here is how you could do it.
```{r}
# let's use the res_all object from the 08_DE_limma exercise:
res_all_limma <- 

res_all %>%
  decideTests(p.value = 0.05, lfc = 0) %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(-gene, names_to = "contrast", values_to = "DE_direction") %>%
  group_by(contrast) %>%
  summarise(
    upregulated = sum(DE_direction == 1),
    downregulated = sum(DE_direction == -1)
  ) %>%
  pivot_longer(-contrast, names_to = "DE_direction", values_to = "n_genes") %>%
  ggplot(aes(x = contrast, y = n_genes, fill = DE_direction)) +
  geom_col(position = "dodge") +
  theme_bw() +
  coord_flip() +
  geom_text(aes(label = n_genes),
            position = position_dodge(width = .9),
            hjust = "inward") +
  labs(y="Number of DE genes") +
  ggtitle("Summary of DE genes by contrast")
```

If we want to show the same amount of information, in a more informative way, a venn diagram is often a better alternative. Here's an easy way to get that visualization if you use either edgeR or limma for your analysis.

```{r venn-allContrasts, fig.height=7, fig.width=10}
# same as before, we can make the plot from the decideTests output
res_all %>%
  decideTests(p.value = 0.01, lfc = 0) %>%
  vennDiagram(include=c("up", "down"),
              lwd=0.75,
              mar=rep(2,4), # increase margin size
              counts.col= c("red", "blue"),
              show.include=TRUE)
```

