---
title: "Motif Analysis"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output:
  html_document:
    code_folding: show
    embed-resources: true
editor_options:
  markdown:
    wrap: 72
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Getting Things Ready

As usual, make sure we have the right packages for this exercise

```{r ready-packages}
if (!require("pacman")) install.packages("pacman"); library(pacman)

# let's load all of the files we were using and want to have again today
p_load("tidyverse", "knitr", "readr",
       "pander", "BiocManager", 
       "dplyr", "stringr", 
       "purrr", # for working with lists (beautify column names)
       "reactable" # for pretty tables.
       ) 

# We also need these Bioconductor packages today.
# p_load("edgeR",
#        "AnnotationDbi", "org.Sc.sgd.db",
#        "pathview","clusterProfiler", "ggupset",
#        "KEGGgraph", "ggkegg", "patchwork",
#        "igraph", "tidygraph", "ggfx")

p_load("biomaRt")
```

# Description

In this class exercise, we will explore the use of the MEME suite for motif analysis.

# Learning outcomes

At the end of this exercise, you should be able to:

-  Retrieve upstream sequences for desired genes in organisms present in the Ensembl database.
-  


# Loading in the `DE_yeast_TF_stress.txt` gene file we used in clustering.

```{r load-res}


# if (!exists("res", envir = .GlobalEnv)) {
#   # If variable doesn't exist, load from
#   url <- "https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/yeast_res_all_edgeR.Rds"
#   res <- read_rds(url)
#   # assign("res", data, envir = .GlobalEnv)
#   cat(paste("Loaded res from", url, "\n"))
#   rm("url", envir = .GlobalEnv)
# } else {
#   url <- "https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/yeast_res_all_edgeR.Rds"
#   cat(paste("res object is already loaded. Skipping loading from", url, "\n"))
#   rm("url", envir = .GlobalEnv)
# }

# let's also get the res object for all of the contrasts at once, (from GitHub)
res_all <- readr::read_tsv("https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/DE_yeast_TF_stress.txt.gz")
```

# Loading in genome database

First, let's make sure the library is loaded and print the version for
reproducibility:

```{r}
library(biomaRt)
packageVersion("biomaRt")
```

## Selecting an Ensembl BioMart database and dataset

```{r}
ensembl <- useEnsembl(biomart = "genes")
ensembl
```

Not we have loaded the "gene" database, but haven't selected any dataset
yet. Let's find the one we want:

```{r}
searchDatasets(mart = ensembl, 
               pattern = "scerevisiae")
```

So the dataset needs to be "scerevisiae_gene_ensembl" for yeast. Can you
find it for your organism?

```{r}
# re-assign ensembl with the dataset we want.
ensembl <- useEnsembl(biomart = "genes", 
                      dataset = "scerevisiae_gene_ensembl")
```

```{r}
ORFs_in_analysis = 
example <- getSequence(id = ORFs_in_analysis,
                 type="ensembl_gene_id",
                 seqType="coding_gene_flank",
                 upstream=500, 
                 mart=ensembl)
```

thoughts on the exercise: Example 1 EtOH MSN2/4 that Jeff had in slack
is best one

Example 2 for salt, skn7 is the top one.

I should do an edgeR treat with strict cutoffs (fdr\<0.01, logFC\>2)

for Eukaryotic memes, change motif lengths minimum to 5. (for msn),
change length max to 20. importance of combinatorial & probabilistic
binding in eukaryotic systems. I will put the default numbers in, and
let Jeff change it during his explanation


# Questions

1.  Can you explain the concept of gene set enrichment analysis and its
    relevance in functional genomics?

2.  How would you interpret the results obtained from clusterProfiler,
    specifically focusing on enriched terms and associated p-values?

3.  Compare and contrast KEGG enrichment with other tools or methods
    used for functional enrichment analysis.

4.  Can you find the KEGG organism for the organism that you are
    interested in?

Be sure to knit this file into a pdf or html file once you're finished.

System information for reproducibility:

```{r session-info}
pander::pander(sessionInfo())
```

