---
title: "Motif Analysis"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output:
  html_document:
    code_folding: show
    embed-resources: true
editor_options:
  markdown:
    wrap: 72
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Getting Things Ready

As usual, make sure we have the right packages for this exercise

```{r ready-packages}
if (!require("pacman")) install.packages("pacman"); library(pacman)

# let's load all of the files we were using and want to have again today
p_load("tidyverse", "knitr", "readr",
       "pander", "BiocManager", 
       "dplyr", "stringr", 
       "purrr", # for working with lists (beautify column names)
       "reactable" # for pretty tables.
       ) 

# We also need these Bioconductor packages today.
# p_load("edgeR",
#        "AnnotationDbi", "org.Sc.sgd.db",
#        "pathview","clusterProfiler", "ggupset",
#        "KEGGgraph", "ggkegg", "patchwork",
#        "igraph", "tidygraph", "ggfx")

p_load("biomaRt")

p_load("memes")
p_load("Biostrings")

library(biomaRt)
library(memes)
```

### We need to install MEME software on your computer if you don't already have it.

This code only downloads & installs the software if not on your computer. The output when installing is VERY long, you can close the output by clicking the small x in the top right corner of the output box.

IF YOU ARE USING LAB MAC for this exercise, or if you don't have XCode on your computer, you'll need to download XCode for mac `xcode-select --install`
```{bash}
xcode-select --install
```


```{bash}
# latest version as of 20 Oct 2023, 
# change to newer version as needed
version=5.5.4

if ! command -v $HOME/meme/bin/meme &> /dev/null; then
  curl -o meme-$version.tar.gz https://meme-suite.org/meme/meme-software/$version/meme-$version.tar.gz
  tar zxf meme-$version.tar.gz
  cd meme-$version
  ./configure --prefix=$HOME/meme --with-url=http://meme-suite.org/ --enable-build-libxml2 --enable-build-libxslt
  make
  make test
  make install
fi
```

Ensure that your download worked:
```{r}
# quick TRUE/FALSE
check_meme_install()
# More details of the install
meme_is_installed()
```

If install didn't work, troubleshooting is needed.

# Description

In this class exercise, we will explore the use of the MEME suite for motif analysis.

# Learning outcomes

At the end of this exercise, you should be able to:

-  Retrieve upstream sequences for desired genes in organisms present in the Ensembl database.
-  


# Loading in the `DE_yeast_TF_stress.txt` gene file we used in clustering.

```{r load-FClist}
FC_list <-
  readr::read_tsv(
    "https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/DE_yeast_TF_stress.txt.gz",
    # make those names less painful to type:
    name_repair = "universal"
  ) 
```

# Loading in genome database

First, let's make sure the library is loaded and print the version for
reproducibility:

```{r}
library(biomaRt)
packageVersion("biomaRt")
```

## Selecting an Ensembl BioMart database and dataset

```{r}
ensembl <- useEnsembl(biomart = "genes")
ensembl
```

Not we have loaded the "gene" database, but haven't selected any dataset
yet. Let's find the one we want:

```{r}
searchDatasets(mart = ensembl, 
               pattern = "scerevisiae")
```

So the dataset needs to be "scerevisiae_gene_ensembl" for yeast. Can you
find it for your organism?

```{r}
# re-assign ensembl with the dataset we want.
ensembl <- useEnsembl(biomart = "genes", 
                      dataset = "scerevisiae_gene_ensembl")
```

```{r}
ORFs_in_analysis = FC_list$ID
seq <- getSequence(id = ORFs_in_analysis,
                 type="ensembl_gene_id",
                 seqType="coding_gene_flank",
                 upstream=500, 
                 mart=ensembl,
                 useCache = TRUE)

# filter out "sequence unavailable" rows
seq <- seq |> filter(coding_gene_flank != "Sequence unavailable")
```

Now we have the 500bp upstream sequences for all of the genes in the genome.
Note that for yeast, this isn't too much data for our computers. If you're working with a larger genome, you might need to subset the gene list down just to the genes in your gene list.


# Get lists of genes that we want to run motif analysis on.


What gene list do you think has the most interesting biological meaning for motif analysis? One comparison to consider is the genes that are most different in the EtOH response between WT and Msn2/4∆∆ samples. Let's get that list of genes with very negative FCs now:
```{r}
msn24_EtOH_down <- FC_list |>
  filter(logFC..WT.v.msn24.mutant..EtOH.response < -2) |>
  filter(FDR..WT.v.msn24.mutant..EtOH.response<0.01) |>
  dplyr::select(ID) %>%
  # add the upstream seqs as a new column
  left_join(seq, by=c("ID" = "ensembl_gene_id")) |>
  drop_na("coding_gene_flank")

# let's take a quick look at the genes identified.
glimpse(msn24_EtOH_down)
```
Perfect, but we need the data to be formatted like a fasta file. Like this:
```{r}
# create Biostrings object
msn24_EtOH_down_fa <- DNAStringSet(msn24_EtOH_down$coding_gene_flank)
# add gene names
names(msn24_EtOH_down_fa) <- msn24_EtOH_down$ID
```


## Run MEME enrichment on those sequences.
```{r}
meme_msn24_EtOH_down <- runMeme(msn24_EtOH_down_fa, 
        minw = 8, # default is 8
        maxw= 50, #default is 50
        parse_genomic_coord=FALSE,
        silent=F
        )

meme_msn24_EtOH_down
```


## Run streme

A newer alternative function, part of the MEME suite, that is better suited to looking for shorter motifs which are common in eukaryotes. 


We can use the promoter sequences from the entire genome as a background to model the null distribution, let's create a background sequences object too for control.
```{r}
# create a DNAStringSet object from our FC_list we created above.
background_fa <- DNAStringSet(left_join(FC_list,seq, by=c("ID" = "ensembl_gene_id")) |>
                                       drop_na("coding_gene_flank") |> pull(coding_gene_flank))
# add gene names
names(background_fa) <- left_join(FC_list,seq, by=c("ID" = "ensembl_gene_id")) |>
                                       drop_na("coding_gene_flank") |> pull(ID)
```

```{r}
streme_msn24_EtOH_down <- runStreme(msn24_EtOH_down_fa, 
        control= background_fa,
        # control= "shuffle",
        minw = 5, # default is 8
        maxw= 15, #default is 15
        parse_genomic_coord=FALSE,
        silent=TRUE
        )

streme_msn24_EtOH_down

dreme_msn24_EtOH_down <- runDreme(msn24_EtOH_down_fa, 
        control= background_fa,
        # outdir = "auto"#, 
        parse_genomic_coord=FALSE,
        silent=TRUE#,
        # maxk=9
        # meme_path = NULL, 
        # silent = TRUE, ...
        )

dreme_msn24_EtOH_down
```

  

YEAST/YEASTRACT_20130918.meme

```{r}
meme_tomtom <- runTomTom(input=meme_msn24_EtOH_down,
          database= "~/Downloads/motif_databases/YEAST/YEASTRACT_20130918.meme")

view_motifs(meme_tomtom$best_match_motif)
```
```{r fig.height=8, fig.width=4}
# assign a directory where we can see the output files saved
out_dir <- path.expand("~/Desktop/Genomic_Data_Analysis/Analysis/memes/streme_msn24_EtOH_down")

# run TomTom analysis
streme_tomtom_msn24_EtOH_down <- runTomTom(input=streme_msn24_EtOH_down,
                 # evalue = FALSE,
                 norc=TRUE,
                 thresh=0.5,
                 motif_pseudo=1,
                 database= "~/Downloads/motif_databases/YEAST/YEASTRACT_20130918.meme",
                 outdir = out_dir) #|>
  # pull(tomtom) |>
  # View()
```

Now, the path where you had the file save to & let's see our results. Open tomtom.html that was saved.

```{r}
view_motifs(streme_tomtom_msn24_EtOH_down$best_match_motif,relative_entropy=F, normalise.scores = T, 
            use.type = "ICM", 
            method="WPCC",
            tryRC=F) # Don't show reverse complement matches
```
It can sometimes be helpful to manually look at the alignments to see if there's anything unexpected going on. We can use the command `view_tomtom_hits()` to do this. The figures aren't publication quality, but are useful for us to see.
```{r}
view_tomtom_hits(streme_tomtom_msn24_EtOH_down, top_n = 3)
```

Cool! We see Msn2/4 promoter sequence motif is strongly enriched in the promoter sequences of the genes that are lower in Msn2/4∆∆ vs WT response to EtOH.

Let's do a sanity check, and do the exact same thing but for genes that were *higher* in the same contrast, and see what enrichments if any are found. We can put just the code we need:
```{r}
# create gene list
msn24_EtOH_up <- FC_list |>
  # I want more than 8 genes, so I lowered the FC cutoff
  filter(logFC..WT.v.msn24.mutant..EtOH.response >2) |>
  filter(FDR..WT.v.msn24.mutant..EtOH.response<0.01) |>
  dplyr::select(ID) %>%
  # add the upstream seqs as a new column
  left_join(seq, by=c("ID" = "ensembl_gene_id")) |>
  drop_na("coding_gene_flank")

# create Biostrings object
msn24_EtOH_up_fa <- DNAStringSet(msn24_EtOH_up$coding_gene_flank)
# add gene names
names(msn24_EtOH_up_fa) <- msn24_EtOH_up$ID

# we already have background genes, so don't need to put that code again

streme_msn24_EtOH_up <- runStreme(msn24_EtOH_up_fa, 
        control= background_fa,
        # control= "shuffle",
        minw = 5, # default is 8
        maxw= 15, #default is 15
        parse_genomic_coord=FALSE,
        silent=TRUE
        )

# assign a directory where we can see the output files saved
out_dir <- path.expand("~/Desktop/Genomic_Data_Analysis/Analysis/memes/streme_msn24_EtOH_up")

# run TomTom analysis
streme_tomtom_msn24_EtOH_up <- runTomTom(input=streme_msn24_EtOH_up,
                 # evalue = FALSE,
                 norc=TRUE,
                 thresh=0.5,
                 motif_pseudo=1,
                 database= "~/Downloads/motif_databases/YEAST/YEASTRACT_20130918.meme",
                 outdir = out_dir)
```



thoughts on the exercise: Example 1 EtOH MSN2/4 that Jeff had in slack
is best one

Example 2 for salt, skn7 is the top one.

I should do an edgeR treat with strict cutoffs (fdr\<0.01, logFC\>2)

for Eukaryotic memes, change motif lengths minimum to 5. (for msn),
change length max to 20. importance of combinatorial & probabilistic
binding in eukaryotic systems. I will put the default numbers in, and
let Jeff change it during his explanation


# Questions

1.  

Be sure to knit this file into a pdf or html file once you're finished.

System information for reproducibility:

```{r session-info}
pander::pander(sessionInfo())
```

