---
title: "ChIP-seq Analysis"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output:
  html_document:
    code_folding: show
    embed-resources: true
editor_options:
  markdown:
    wrap: 72
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Description

In this class exercise, we will practice analyzing ChIP-seq data.

## Learning Outcomes

At the end of this exercise, you should be able to:

-   

## Exercise Background

In this exercise, we will look at the only published ChIP-seq data for
Msn2/4 mutants under EtOH stress conditions similar to those we have
been exploring throughout this course. The data comes from []. We will
go through this analysis together, starting from the raw fastq files.

## Prepare Software

We will need to prepare our computing environment for this analysis.
This analysis will use both R and terminal based commands.

### R packages

```{r ready-packages}
# Ensure required packages are installed
if (!require("pacman")) install.packages("pacman"); library(pacman)

# Load necessary packages
# p_load("tidyverse", "knitr", "readr", "pander", "BiocManager", 
#        "dplyr", "stringr", #"purrr", "reactable",
#        "biomaRt", "memes", "Biostrings", "curl", "universalmotif")

# library(biomaRt)
# library(memes)
# library(curl)
# library(universalmotif)
library(tidyverse)
library(ShortRead) # added by me, using rockerfeller approach.
library(ChIPQC)
library(ChIPseeker)
library(DiffBind)
library(clusterProfiler)
library(AnnotationDbi)
```

### Conda environments

Similar to exercise 03_Working_with_Sequences.Rmd, we need to use
softwares that aren't available in R. One way to do that is the way
shown here, creating and activating conda environments. If you haven't
done set-up of conda on your system as described in exercise 03, this
code likely won't run for you.

#### FastQC

```{bash create-QCcondaEnvironment, engine.opts='-l'}
# create an enviornment for our QC packages
if conda info --envs | grep -q QC; then echo "environment 'QC' already exists"; else conda create -y -n QC fastqc multiqc; fi

# see available conda environments
conda env list

# activate our QC environment
conda activate QC

# make sure desired packages are working
which fastqc
which multiqc

# get the versions of each software
fastqc -v
multiqc --version

# it's always good coding practice to deactivate a conda environment at the end of a chunk
conda deactivate
```

#### MACS2

A popular software for peak-calling ChIP-seq data is
[MACS2](https://pypi.org/project/MACS2/). It uses a Poisson distribution-based model to background identify significant peaks in ChIP-seq data.

```{bash create-MACS2condaEnvironment, engine.opts='-l'}
# create an enviornment for our QC packages
if conda info --envs | grep -q MACS; then echo "environment 'MACS' already exists"; else conda create -y -n MACS macs2 -c bioconda; fi

# see available conda environments
conda env list

# activate our QC environment
conda activate MACS

# make sure desired packages are working
which fastqc
which multiqc

# get the versions of each software
fastqc -v
multiqc --version

# it's always good coding practice to deactivate a conda environment at the end of a chunk
conda deactivate
```



## Data Retreival

### fastq files

### Reference genome

## fastQC

## trimming

## create BAM files




# Analysis: Motif Discovery for *msn2/4*∆∆ vs WT Response to EtOH

## Retrieve LogFC and FDR values

Let's load in the `DE_yeast_TF_stress.txt` file containing logFC and FDR
values for a variety of yeast strains and stress conditions.

```{r load-FClist}
# Load gene file used in clustering
FC_list <- readr::read_tsv(
  "https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/DE_yeast_TF_stress.txt.gz",
  name_repair = "universal"
) 
```

## Selecting an Ensembl BioMart database and dataset

```{r}
ensembl <- useEnsembl(biomart = "genes")
ensembl
```

Not we have loaded the "gene" database, but haven't selected a dataset
yet. Let's find the one we want:

```{r}
searchDatasets(mart = ensembl, 
               pattern = "scerevisiae")
```

So the dataset needs to be "scerevisiae_gene_ensembl" for yeast. Can you
find it for your organism?

```{r}
# Assign ensembl with the desired dataset
ensembl <- useEnsembl(biomart = "genes", 
                      dataset = "scerevisiae_gene_ensembl")
```

## Retrieve Upstream Sequences

```{r}
# Retrieve the ORF IDs included in the dataset
ORFs_in_analysis = FC_list$ID

# Get the upstream sequences for each gene and save in data.frame
seq <- getSequence(id = ORFs_in_analysis,
                 type="ensembl_gene_id",
                 seqType="coding_gene_flank",
                 upstream=500, 
                 mart=ensembl,
                 useCache = TRUE)

# Filter out rows with no sequence found
seq <- seq |> filter(coding_gene_flank != "Sequence unavailable")

# Display obtained sequences
glimpse(seq)
```

Now we have the 500bp upstream sequences for (almost) all of the genes
in the genome. Note that for yeast, this isn't too much data for a
laptop. If you're working with a larger genome, you might need to subset
the gene list down just to the genes in your gene list.

# Motif Analysis for Genes Downregulated in EtOH Response

In order to continue, we need a list of genes.

What gene list do you think has the most interesting biological meaning
for motif analysis? One comparison to consider is the genes that are
most different in the EtOH response between WT and *msn2/4*∆∆ samples.
Let's get that list of genes with large magnitude and statistically
significant negative logFC values now:

```{r}
# Create gene list for downregulated genes
msn24_EtOH_down <- FC_list |>
  dplyr::filter(logFC..WT.v.msn24.mutant..EtOH.response < -2) |>
  dplyr::filter(FDR..WT.v.msn24.mutant..EtOH.response<0.01) |>
  dplyr::select(ID) %>%
  # add the upstream sequences as a new column
  left_join(seq, by=c("ID" = "ensembl_gene_id")) |>
  drop_na("coding_gene_flank")

# Glimpse at the genes identified.
glimpse(msn24_EtOH_down)
```

For MEME analysis, the gene list needs to be formatted like a fasta
file. Here is how we can do that:

```{r}
# create Biostrings object
msn24_EtOH_down_fa <- Biostrings::DNAStringSet(msn24_EtOH_down$coding_gene_flank)
# Add gene names
names(msn24_EtOH_down_fa) <- msn24_EtOH_down$ID
```

Let's create a folder to which we can save output files.

```{r}
# Choose output directory for the output files to be saved
out_dir <- path.expand("~/Desktop/Genomic_Data_Analysis/Analysis/memes/")

# Create out_dir directory if doesn't already exist
if (!dir.exists(out_dir)) {
  dir.create(out_dir, recursive = TRUE)
}
```

## Run Meme enrichment

This was the first algorithm developed in the MEME-suite, and is still
widely used. However, it is not always the best approach. MEME is
recommended when you have fewer than 50 sequences, while STREME is
recommended when you have more. Also, the default parameter settings are
often **not** the best options depending on your organism (e.g, bacteria
tend to have longer motifs). We can adjust these settings in class to
parameters better suited to our data (based on biological domain
knowledge). See all of the parameter options at:
<https://meme-suite.org/meme/doc/meme.html>

```{r}
# Run Meme
meme_msn24_EtOH_down <- runMeme(msn24_EtOH_down_fa, 
        minw = 8, # default is 8, for yeast I use 5
        maxw= 50, #default is 50, for yeast I use 20
        mod= "zoops", #zero or one occurence per sequence
        parse_genomic_coord=FALSE,
        silent=F,
        outdir = path.expand(paste0(out_dir, "meme_msn24_EtOH_down"))
        )

# Display Meme results
meme_msn24_EtOH_down
```

## Run STEME enrichment

STEME is a newer MEME suite tool, that is better suited to looking for
shorter motifs which are common in Eukaryotes.

For `runStreme`, properly setting the `control` parameter is key to
discovering biologically relevant motifs. Often, using
`control = "shuffle"` will produce a suboptimal set of motifs; however,
some discriminative analysis designs don't have proper "control" regions
other than to shuffle.

For our analysis, we can use the promoter sequences from the entire
genome as a background to model the null distribution, let's create an
object in R with all of the 500bp upstream sequences from `seq` to use
as a control.

```{r}
# create a DNAStringSet object from our FC_list we created above.
background_fa <- DNAStringSet(left_join(FC_list,seq, by=c("ID" = "ensembl_gene_id")) |>
                                       drop_na("coding_gene_flank") |> pull(coding_gene_flank))
# add gene names
names(background_fa) <- left_join(FC_list,seq, by=c("ID" = "ensembl_gene_id")) |>
                                       drop_na("coding_gene_flank") |> pull(ID)
```

```{r}
# Run Streme
streme_msn24_EtOH_down <- runStreme(msn24_EtOH_down_fa, 
        control= background_fa,
        minw = 8, # default is 8, for yeast I use 5
        maxw= 15, #default is 15, for yeast i use 20
        parse_genomic_coord=FALSE,
        silent=TRUE,
        outdir = path.expand(paste0(out_dir, "streme_msn24_EtOH_down"))
        )

# Display Streme results
streme_msn24_EtOH_down
```

## Run TomTom analysis

Now, we can see if any of those identified motifs correspond to known
motifs in databases. The TomTom algorithm, part of the MEME-suite,
allows us to just that, comparing our identified motifs to known motifs
in annotation databases.

First, we need to download the YEASTRACT database file called
(YEASTRACT_20130918.meme). We will pull this file from Github, but you
can see and download all of the available databases at:
<https://meme-suite.org/meme/doc/download.html>.

```{r}
# Download YEASTRACT database .meme

# Define URL where data is located
URL_to_download <-
  "https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/reference/YEAST/YEASTRACT_20130918.meme"

# Choose desired destination for the reference database file
db_destination <-
  path.expand("~/Desktop/Genomic_Data_Analysis/Reference/YEASTRACT_20130918.meme")

# Download the file and save to db_destination
curl::curl_download(
  url = URL_to_download,
  destfile = db_destination,
  quiet = FALSE,
  mode = "wb"
)
```

```{r}
# Run TomTom on motifs found by runMeme()
meme_tomtom_msn24_EtOH_down <-
  runTomTom(
    input = meme_msn24_EtOH_down,
    norc = TRUE,
    thresh = 10,
    motif_pseudo = 0.1,
    database = db_destination,
    outdir = path.expand(paste0(out_dir, "tomtom_meme_msn24_EtOH_down"))
  )

# View Meme TomTom results
view_motifs(meme_tomtom_msn24_EtOH_down$best_match_motif)
```

Go to the `out_dir` directory on your computer, and open the tomtom.html
file. How convincing are those matches? With the default options, I'm
not impressed.

Let's run TomTom on the output of the `runStreme()` command above, and
see what we find.

```{r}
# Run TomTom on motifs found by runStreme()
streme_tomtom_msn24_EtOH_down <-
  runTomTom(
    input = streme_msn24_EtOH_down,
    norc = TRUE,
    thresh = 10,
    motif_pseudo = 0.1,
    database = db_destination,
    outdir = path.expand(paste0(out_dir, "tomtom_streme_msn24_EtOH_down"))
  )

# View Streme TomTom results
view_motifs(streme_tomtom_msn24_EtOH_down$best_match_motif,
            relative_entropy = FALSE, 
            normalise.scores = TRUE, 
            use.type = "ICM", 
            method = "WPCC",
            tryRC = FALSE)
```

It can sometimes be helpful to manually look at the alignments to see if
there's anything unexpected going on. We can use the command
`view_tomtom_hits()` to do this. The figures aren't publication quality,
but can be useful to see.

```{r}
view_tomtom_hits(streme_tomtom_msn24_EtOH_down, top_n = 3)
```

Cool! We see Msn2/4 promoter sequence motif is strongly enriched in the
promoter sequences of the genes that are lower in Msn2/4∆∆ vs WT
response to EtOH.

Let's do a sanity check, and do the exact same thing but for genes that
were *higher* in the same contrast, and see what enrichments if any are
found. We can put just the code we need:

```{r motif-Msn24EtOH_up}
# Create gene list
msn24_EtOH_up <- FC_list |>
  # I want more than 8 genes, so I lowered the FC cutoff
  filter(logFC..WT.v.msn24.mutant..EtOH.response > 1) |>
  filter(FDR..WT.v.msn24.mutant..EtOH.response<0.01) |>
  dplyr::select(ID) %>%
  # add the upstream seqs as a new column
  left_join(seq, by=c("ID" = "ensembl_gene_id")) |>
  drop_na("coding_gene_flank")

# create Biostrings object
msn24_EtOH_up_fa <- DNAStringSet(msn24_EtOH_up$coding_gene_flank)
# add gene names
names(msn24_EtOH_up_fa) <- msn24_EtOH_up$ID


# we already have background genes, so don't need to put that code again...

# Run Streme
streme_msn24_EtOH_up <- runStreme(msn24_EtOH_up_fa, 
        control= background_fa,
        # control= "shuffle",
        minw = 5, # default is 8, for yeast I use 5
        maxw= 20, #default is 15, for yeast I use 20
        parse_genomic_coord=FALSE,
        silent=TRUE,
  outdir = path.expand(paste0(out_dir, "streme_msn24_EtOH_up"))
        )

# Run TomTom analysis
streme_tomtom_msn24_EtOH_up <- runTomTom(
  input = streme_msn24_EtOH_up,
  norc = TRUE,
  thresh = 10,
  motif_pseudo = 0.1,
  database = db_destination,
  outdir = path.expand(paste0(out_dir, "tomtom_streme_msn24_EtOH_up"))
  )
```

Are Msn2/4 motifs enriched here? What biological meaning can we take
away from this?

# *skn7*∆ exposed to salt.

Now, let's take a look another mutant from this study exposed to a
different stressor.

```{r motif-skn7NaCl}
# create gene list
skn7_NaCl_down <- FC_list |>
  filter(logFC..WT.v.skn7.mutant..NaCl.response < -2) |>
  filter(FDR..WT.v.skn7.mutant.NaCl.response<0.00001) |>
  dplyr::select(ID) %>%
  # add the upstream seqs as a new column
  left_join(seq, by=c("ID" = "ensembl_gene_id")) |>
  drop_na("coding_gene_flank")

# create Biostrings object
skn7_NaCl_down_fa <- DNAStringSet(skn7_NaCl_down$coding_gene_flank)
# add gene names
names(skn7_NaCl_down_fa) <- skn7_NaCl_down$ID

# we already have background genes, so don't need to put that code again...

# Run Streme
streme_skn7_NaCl_down <- runStreme(skn7_NaCl_down_fa, 
        control= background_fa,
        # control= "shuffle",
        minw = 8, # default is 8, for yeast I use 5
        maxw= 15, #default is 15, for yeast I use 20
        evalue=TRUE,
        parse_genomic_coord=FALSE,
        silent=TRUE,
        outdir = path.expand(paste0(out_dir, "streme_skn7_NaCl_down"))
        )

# Run TomTom analysis
streme_tomtom_skn7_NaCl_down <- runTomTom(
  input = streme_skn7_NaCl_down,
  # norc = TRUE,
  thresh = 20,
  motif_pseudo = 0.1,
  database = db_destination,
  outdir = path.expand(paste0(out_dir, "tomtom_streme_skn7_NaCl_down"))
  )

```

Notice we have fewer genes this time, let's try meme instead

```{r}
# Run Meme
meme_skn7_NaCl_down <- runMeme(
  skn7_NaCl_down_fa,
  # control= "shuffle", #background_fa,
  # objfun="de",
  parse_genomic_coord = FALSE,
  minw = 5,
  maxw = 20,
  markov_order = 2,
  mod= "zoops", #zero or one occurence per sequence
  seed = 0,
  dna = T,
  revcomp = T,
  evt = 0.1,
  outdir = path.expand(paste0(out_dir, "meme_skn7_NaCl_down"))
)

# Run TomTom analysis
meme_tomtom_skn7_NaCl_down <- runTomTom(
  input = meme_skn7_NaCl_down,
  # norc = TRUE,
  thresh = 10,
  motif_pseudo = 0.1,
  min_overlap = 5,
  database = db_destination,
  outdir = path.expand(paste0(out_dir, "tomtom_meme_skn7_NaCl_down"))
  )
```

# Questions

1.  What is the difference in motifs when you change the settings for
    analyzing the msn2/4 mutant vs wild-type ethanol response using
    "zoops" or "anr" as the setting?

2.  Perform MEME and STREME on the wild-type salt response and wild-type
    ethanol response (using a log2 FC of 3 and FDR \< 0.01) to identify
    motifs in genes that are strongly induced for each stress, and then
    follow that analysis with TOMTOM. What TFs may be shared and what
    TFs may be different between the two stress responses?

Be sure to knit this file into a pdf or html file once you're finished.

System information for reproducibility:

```{r session-info}
pander::pander(sessionInfo())
```
