---
title: "Differential Expression Analysis with DeSeq2"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output: html_document
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
set.seed("1492")
```

# Getting Things Setup

As usual, make sure we have the right packages for this exercise

```{r ready-packages, results="hide",message=FALSE}
if (!require("pacman")) install.packages("pacman"); library(pacman)

# let's load all of the files we were using and want to have again today
p_load("tidyverse", "knitr", "readr",
       "pander", "BiocManager", 
       "dplyr", "stringr", 
       "purrr", # for working with lists (beautify column names)
       "reactable") # for pretty tables.

# We also need these Bioconductor packages today.
p_load("DESeq2", "AnnotationDbi", "org.Sc.sgd.db")
```

# Description

This will be our second differential expression analysis workflow, converting gene counts across samples into meaningful information about genes that appear to be significantly differentially expressed between samples. This is inspired heavily by: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html. 

# Learning outcomes

At the end of this exercise, you should be able to:

-   Utilize the DESeq2 package to identify differentially expressed genes.


```{r load-libraries}
library(DESeq2)
library(org.Sc.sgd.db)
# for ease of use, set max number of digits after decimal
options(digits=3)
```

# Loading in the featureCounts object 

We saved this file at the end the exercise (Read_Counting.Rmd). Now we can load that object back in and assign it to the variable fc. Be sure to change the file path if you have saved it in a different location. This is the same way we started the edgeR analysis.

```{r load-fc, error=TRUE}
path_fc_object <- path.expand("~/Desktop/Genomic_Data_Analysis/Data/yeast_fc_output.Rds")
fc <- readRDS(file = path_fc_object)
```


If you don't have that file for any reason, the below code chunk will load a copy of it from Github.
```{r altLoad-fc}
if( !exists("fc") )
{
  fc <- read_rds('https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/ethanol_stress/yeast_fc_output.Rds')
}
```

We will create the data frame again that has all of the metadata information. 

```{r generate-metadata}
sample_metadata <- tribble(
  ~Sample,                      ~Genotype,    ~Condition,
  "YPS606_MSN24_ETOH_REP1_R1",   "msn24dd",   "EtOH",
  "YPS606_MSN24_ETOH_REP2_R1",   "msn24dd",   "EtOH",
  "YPS606_MSN24_ETOH_REP3_R1",   "msn24dd",   "EtOH",
  "YPS606_MSN24_ETOH_REP4_R1",   "msn24dd",   "EtOH",
  "YPS606_MSN24_MOCK_REP1_R1",   "msn24dd",   "unstressed",
  "YPS606_MSN24_MOCK_REP2_R1",   "msn24dd",   "unstressed",
  "YPS606_MSN24_MOCK_REP3_R1",   "msn24dd",   "unstressed",
  "YPS606_MSN24_MOCK_REP4_R1",   "msn24dd",   "unstressed",
  "YPS606_WT_ETOH_REP1_R1",      "WT",        "EtOH",
  "YPS606_WT_ETOH_REP2_R1",      "WT",        "EtOH",
  "YPS606_WT_ETOH_REP3_R1",      "WT",        "EtOH",
  "YPS606_WT_ETOH_REP4_R1",      "WT",        "EtOH",
  "YPS606_WT_MOCK_REP1_R1",      "WT",        "unstressed",
  "YPS606_WT_MOCK_REP2_R1",      "WT",        "unstressed",
  "YPS606_WT_MOCK_REP3_R1",      "WT",        "unstressed",
  "YPS606_WT_MOCK_REP4_R1",      "WT",        "unstressed") %>%
  # make Condition and Genotype a factor (with baseline as first level) for DESeq2
  mutate(
    Genotype = factor(Genotype,
                      levels = c("WT", "msn24dd")),
    Condition = factor(Condition,
                       levels = c("unstressed", "EtOH"))
  ) #%>%
  # arrange(desc(Condition)) %>%
  # arrange(desc(Genotype))
```


# Count loading and Annotation

The count matrix is used to construct a DESeqDataSet class object. This is the main data class in the DESeq2 package. The DESeqDataSet object is used to store all the information required to fit a generalized linear model to the data, including library sizes and dispersion estimates as well as counts for each gene.

Because we used the featureCounts function (Liao, Smyth, and Shi 2013) in the Rsubread package, the matrix of read counts can be directly provided from the `"counts"` element in the list output. The count matrix and column data can typically be read into R from flat files using base R functions such as read.csv or read.delim.

With the count matrix, cts, and the sample information, coldata, we can construct a DESeqDataSet:

```{r load-counts}
# notice the different design specification
dds <- DESeqDataSetFromMatrix(countData = fc$counts,
                              colData = sample_metadata,
                              design = ~ 1 + Genotype + Condition + Genotype:Condition)

# simplify the column names to make them pretty
colnames(dds) <- str_split_fixed(colnames(dds), "\\.", n = 2)[, 1]

# take a look at the dds object
dds

# compare this to the edgeR process below:
# y <- DGEList(fc$counts, group=group)
# colnames(y) <- sample_metadata$GEOAccession
# y
```

# Filtering to remove low counts


While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful: by removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of count modeling within DESeq2. It can also improve visualizations, as features with no information for differential expression are not plotted in dispersion plots or MA-plots.

Here we perform pre-filtering to keep only rows that have a count of at least 10 for a minimal number of samples. The count of 10 is a reasonable choice for bulk RNA-seq. A recommendation for the minimal number of samples is to specify the smallest group size, e.g. here there are 4 treated samples. If there are not discrete groups, one can use the minimal number of samples where non-zero counts would be considered interesting. One can also omit this step entirely and just rely on the independent filtering procedures available in results(), either IHW or genefilter. See independent filtering section.

```{r filter-lowCounts}
smallestGroupSize <- 4
keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize
dds <- dds[keep,]

# Equivalent version in edgeR:
# keep <- rowSums(cpm(y) > 60) >= 4
# y <- y[keep,]
# summary(keep)
```




# Testing for differential expression

The standard differential expression analysis steps are wrapped into a single function, DESeq. The estimation steps performed by this function are described below, in the manual page for `?DESeq` and in the Methods section of the DESeq2 publication (Love, Huber, and Anders 2014).

Results tables are generated using the function `results`, which extracts a results table with log2 fold changes, p values and adjusted p values. With no additional arguments to `results`, the log2 fold change and Wald test p value will be for the last variable in the design formula, and if this is a factor, the comparison will be the *last level* of this variable over the *reference level* However, the order of the variables of the design do not matter so long as the user specifies the comparison to build a results table for, using the name or contrast arguments of results.

Details about the comparison are printed to the console, directly above the results table. The text, condition treated vs untreated, tells you that the estimates are of the logarithmic fold change log2(treated/untreated).

```{r run-DESeq}
# Now that we have a DESeq2 object, we can can perform differential expression.
dds <- DESeq(dds)
resultsNames(dds)
```

```{r design-contrasts}
# create a model matrix
mod_mat <- model.matrix(design(dds), colData(dds)) 

# define coefficient vectors for each group
WT_MOCK <- colMeans(mod_mat[dds$Genotype == "WT" & dds$Condition == "unstressed", ])
WT_EtOH <- colMeans(mod_mat[dds$Genotype == "WT" & dds$Condition == "EtOH", ])
MSN24_MOCK <- colMeans(mod_mat[dds$Genotype == "msn24dd" & dds$Condition == "unstressed", ])
MSN24dd_EtOH <- colMeans(mod_mat[dds$Genotype == "msn24dd" & dds$Condition == "EtOH", ])
```

The nice thing about this approach is that we do not need to worry about any of this, the weights come from our `colMeans()` call automatically. And now, any contrasts that we make will take these weights into account:

```{r get-results}
res <- results(dds)
res
```

We could have equivalently produced this results table with the following more specific command. Because Genotypemsn24dd:ConditionEtOH is the last variable in the design, we could optionally leave off the contrast argument to extract the comparison of the two levels of Genotypemsn24dd:ConditionEtOH.

```{r}
res <- results(dds, 
               contrast = (MSN24dd_EtOH - MSN24_MOCK) - (WT_EtOH - WT_MOCK)
               )
```


```{r create-resultsTable}
res %>% 
  data.frame() %>%
  rownames_to_column("ORF") %>%
  # add the gene names
  left_join(AnnotationDbi::select(org.Sc.sgd.db,keys=.$ORF,columns="GENENAME"),by="ORF") %>%
  relocate(GENENAME, .after = ORF) %>%
  arrange(padj) %>%
  mutate(log2FoldChange = round(log2FoldChange, 2)) %>%
  mutate(across(where(is.numeric), signif, 3)) %>%
  reactable()

# filter based on padj and a lfc cutoff
res_sig <- subset(res, padj<.01)
res_lfc <- subset(res_sig, abs(log2FoldChange) > 1)
```

```{r}
# let's compare the summaries before and after setting a lfc cutoff:
summary(res, alpha=0.01)
summary(res_lfc, alpha=0.01)

head(res_lfc)
```
Let's take a quick look at the differential expression
```{r}
plotMA(res, alpha=0.01)
```
Plot an individual gene:
```{r}
gene <- "YKL035W"

# Here is the default visualization. Depending on screen size, the xlab 
# might not show all of the groups.
plotCounts(dds, gene="YEL039C", intgroup=c("Genotype","Condition"),
           xlab="Genotype:Condition")

# Make the plot prettier with ggplot(). Note the returnData=TRUE let's us do this.
plotCounts(dds, gene=gene, intgroup=c("Genotype","Condition"),
           xlab="Genotype:Condition", returnData = TRUE) %>%
  rownames_to_column("Sample") %>%
  ggplot(aes(x=Genotype, y=count, color=Condition, shape=Condition)) +
  geom_dotplot(binaxis = "y", stackdir = "center", dotsize=0.75,
               position=position_dodge(0.4), # this seperates by Condition a bit
               fill=NA) +
  labs(x="Genotype",
       y="normalized count",
       title=paste0("Visualizing the expression of ORF: ", gene)
       ) +
  # scale_y_log10() +
  theme_classic()
```


We need to make sure and save our output file(s).
```{r save-res}
# Choose topTags destination
dir_output_DESeq2 <-
  path.expand("~/Desktop/Genomic_Data_Analysis/Analysis/DESeq2/")
if (!dir.exists(dir_output_DESeq2)) {
  dir.create(dir_output_DESeq2, recursive = TRUE)
}

# for sharing with others, a tsv for the res output is convenient.
# Depending on what people need, we can save res object as is or beautify it.
res %>% 
  data.frame() %>%
  rownames_to_column("ORF") %>%
  left_join(AnnotationDbi::select(org.Sc.sgd.db,keys=.$ORF,columns="GENENAME"),by="ORF") %>%
  relocate(GENENAME, .after = ORF) %>%
  # arrange(padj) %>%
  # mutate(log2FoldChange = round(log2FoldChange, 2)) %>%
  # mutate(across(where(is.numeric), signif, 3)) %>%
  write_tsv(., file = paste0(dir_output_DESeq2, "yeast_res_DESeq2.tsv"))

# for subsequent analysis, let's save the res object as an R data object.
saveRDS(object = res, file = paste0(dir_output_DESeq2, "yeast_res_DESeq2.Rds"))
```

Be sure to knit this file into a pdf or html file once you're finished.

System information for reproducibility:
```{r}
pander::pander(sessionInfo())
```
