---
title: "Differential Expression Analysis with EdgeR"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output: html_document
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
set.seed("1492")
```

# Getting Things Setup

As usual, make sure we have the right packages for this exercise

```{r ready-packages, results="hide",message=FALSE}

list.of.packages <- c("tidyverse", "here", "knitr", "pander", "BiocManager")

new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages,repos = "http://cran.us.r-project.org")
lapply(list.of.packages, require, character.only = TRUE)

```

We also need the bioconductor package "edgeR" for today's activity.

```{r install-FastqPackages, results="hide", message=FALSE, warning=FALSE}
# make sure we have the CRAN package installed for getting bioconductor packages
if (!requireNamespace("edgeR", quietly = TRUE))
    BiocManager::install("edgeR")

# if (!requireNamespace("ShortRead", quietly = TRUE))
#     BiocManager::install("ShortRead")
```

# Description

This will be our first differential expression analysis workflow, converting gene counts across samples into meaningful information about genes that appear to be significantly differentially expressed between samples

# Learning outcomes

At the end of this exercise, you should be able to:

-   Generate a table of sample metadata .
-   Filter low counts and normalize count data.
-   Utilize the edgeR package to identify differentially expressed genes.


```{r load-libraries, results="hide", message=FALSE, warning=FALSE}
library(edgeR)
```



```{bash download-fq, message=FALSE}
# Be sure to change this file path to the path you want your data to go
DATA_DIR="/Users/$USER/Desktop"
# WARNING: curl doesn't work with relative paths
curl -L -C - -o $DATA_DIR/subset_phenol_heat_rep1.fastq.gz https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/heat_shock/subset_phenol_heat_rep1.fastq.gz?raw=TRUE
```




Be sure to knit this file into a pdf or html file once you're finished.

System information for reproducibility:
```{r}
pander::pander(sessionInfo())
```
