---
title: "Getting Started in R"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output: html_document
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Getting things setup

First things first: Click the "Visual" button in the top-left corner of the code box. This makes the code look more like a word processor. You can always switch back to Source anytime you prefer.

The following code installs a set of R packages used in this document -- if not already installed -- and then loads the packages into R. Note that we utilize the US CRAN repository, but other repositories may be more convenient according to geographic location.

```{r ready-packages, results="hide",message=FALSE}
if (!require("pacman")) install.packages("pacman"); library(pacman)

# the p_load function 
#    A) installs the package if not installed (like install.packages("package_name")),
#    B) loads the package (equivalent of library(package_name))

p_load("tidyverse", # An ecosystem of packages for making life in R easier
       "here", # For locating files easily
       "knitr", # For generating ("knitting") html or pdf files from .Rmd file
       "readr", # For faster and easier reading in files to R
       "pander", # For session info at the end of the document
       "BiocManager", # For installing Bioconductor R packages
       "dplyr" # A key part of the tidyverse ecosystem, has useful functions
       )
```

# Description

This activity is intended to familiarize you with using RStudio and the R ecosystem to analyze genomic data

# Learning outcomes

At the end of this exercise, you should be able to:

-   open, modify, and knit an Rmd file to a pdf/html output
-   relate Rmarkdown to a traditional lab notebook
-   run commands in an Rmarkdown file

# Using R and RStudio

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r using-codeChunk}
# print a statement
print("R code in a .Rmd chunk works just like a script")

# preform basic calculations
2+2
```

R is a useful tool for analyzing data. Let's download a data file from GitHub to work with. First, we will download the file manually and open it. Later, we will download the same file directly from the url.

-   [Click here](https://github.com/clstacy/GenomicDataAnalysis_Fa23/blob/main/data/ethanol_stress/msn2-4_mutants_EtOH.txt) to open the file in GitHub and click the download icon to download it to your computer.

-   Use the "Import Dataset" in the Environment panel of RStudio to open the file browser and select the downloaded file

    -   You'll want to use the "From text (readr)..." option

-   Adjust settings to make sure the file loads in properly.

-   Copy the code that the Import Dataset feature provides for reading in the file and paste it in the code chunk below

```{r load-manually}
# insert here the code used to load the file in from your computer

```

## Load data directly from the URL

Rather than downloading the file manually and then loading it in from where we downloaded it to, we can just load it directly from the URL, as shown below. A word of caution, this won't work with any URL and you can't guarantee the URL will always work in the future.

```{r access-data}
# assign url to a variable
DE_data_url <- "https://raw.githubusercontent.com/clstacy/GenomicDataAnalysis_Fa23/main/data/ethanol_stress/msn2-4_mutants_EtOH.txt"

# download the data from the web
DE_results_msn24_EtOH <-
  read_tsv(file=DE_data_url)

```

Do remember that this function uses the package readr (a part of the tidyverse package we loaded above). If you don't have that package (1) installed and (2) loaded into your script, it won't work. Thankfully, the p_load function takes care of both of these simultaneously.

## Working with data in R

To get a quick summary of our data and how it looks

```{r}
# take a quick look at how the data is structured
glimpse(DE_results_msn24_EtOH)
```

We see in the output there are `r nrow(DE_results_msn24_EtOH)` rows and `r ncol(DE_results_msn24_EtOH)` columns in the data. The same information should be available in the environment panel of RStudio

# Seeing Data in RStudio

If we want to take a closer look at the data, we have a few options. To see just the first few lines we can run the following command:

```{r head-data}
head(DE_results_msn24_EtOH)
```

This can be difficult to look at. For looking at data similar to an Excel file, RStudio allows this by clicking on the name of the data.frame in the top right corner of the IDE. We can also view a file by typing `View(filename)`. To open the data in a new window, click the "pop out" button next to "filter" just above the opened dataset.

## Exploring the data

This dataset includes the log fold changes of gene expression in an experiment testing the ethanol stress response for the YPS606 strain of *S. cerevisiae* and an *msn2/4ΔΔ* mutant. There are also additional columns of metadata about each gene. In later classes, we will cover the details included, but we can already start answering questions.

**Using RStudio, answer the following questions:**

1.  How many genes are included in this study?

2.  Which gene has the highest log fold change in the *msn2/4ΔΔ* mutant EtOH response?

3.  How many HSP genes are differentially expressed (FDR \< 0.01) in unstressed conditions for the mutant?

4.  Do the genes with the largest magnitude fold changes have the smallest p-values?

5.  Which isoform of phosphoglucomutase is upregulated in response to ethanol stress? Do you think *msn2/4* is responsible for this difference?

Be sure to knit this file into a pdf or html file once you're finished.

<br>

System information for reproducibility:

```{r}
pander::pander(sessionInfo())
```
