---
title: "Differential Expression Analysis with EdgeR"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output: html_document
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
set.seed("1492")
```

# Getting Things Setup

As usual, make sure we have the right packages for this exercise

```{r ready-packages}
if (!require("pacman")) install.packages("pacman"); library(pacman)

# let's load all of the files we were using and want to have again today
p_load("tidyverse", "knitr", "readr",
       "pander", "BiocManager", 
       "dplyr", "stringr", 
       "purrr", # for working with lists (beautify column names)
       "reactable") # for pretty tables.

# We also need these Bioconductor packages today.
p_load("edgeR", "AnnotationDbi", "org.Sc.sgd.db", "pathview",
       "clusterProfiler", "KEGGgraph", "ggkegg", "patchwork",
       "igraph", "tidygraph", "ggfx")# for dotplot

# library(ggkegg)
# p_load(ggfx)
# library(igraph)
# library(tidygraph)
# library(dplyr)
# library(patchwork)
```


# Description


In this class exercise, we will explore the use of KEGG pathways in genomic data analysis. KEGG is a valuable resource for understanding biological pathways and functions associated with genomic data.

# Learning outcomes

At the end of this exercise, you should be able to:

-   Understand the basics of KEGG pathways.
-   Learn how to retrieve and analyze pathway information using R.
-   Apply KEGG analysis to genomic data.


```{r load-libraries}
library(edgeR)
library(org.Sc.sgd.db)
# for ease of use, set max number of digits after decimal
options(digits=3)
```

# Loading in the edgeR DE gene file output.


```{r save-res}
# Choose topTags destination
dir_output_edgeR <-
  path.expand("~/Desktop/Genomic_Data_Analysis/Analysis/edgeR/")

# for this, let's load the res objects as an R data object.
res <- read_rds(file = paste0(dir_output_edgeR, "yeast_res_edgeR.Rds"))
res_all <- read_rds(file = paste0(dir_output_edgeR, "yeast_res_all_edgeR.Rds"))
```



# KEGG

First, let's do a simple KEGG enrichment with the limma package, using the function `kegga()`. This approach is nice, because we can directly load in the `res` object we created in the `edgeR` workflow, set our FDR cutoff, and the process happens automatically.

```{r}
# alias2Symbol()

# go <- goana(res, species.KEGG="sce")
# topGO(go, n=10)

# from limma, there is the 
kegg <- kegga(res, species.KEGG="sce", FDR=0.01,
      plot=T)

topKEGG(kegg, sort="up")


gene_data_logFC <- res$table %>% dplyr::select(logFC)

fold_change_geneList <- setNames(object = data.frame(res)$logFC,nm = data.frame(res)$ORF)
head(fold_change_geneList)


if (!requireNamespace("pathview", quietly = TRUE))
    BiocManager::install("pathview")

library("pathview")
test <- pathview(gene.data  = fold_change_geneList,
                     # pathway.id = "sce01100", #metabolic pathways
                     pathway.id = "sce00010", #glycolysis/gluconeogenesis
                     # pathway.id = "sce03050", # proteasome cycle
                     # pathway.id = "sce00020", # TCA cycle
                     # pathway.id = "sce00500", # starch & sucrose metabolism (trehalose)
                     # pathway.id = "sce00030", #PPP
                     species    = "sce",
                     gene.idtype="orf",
                     # expand.node = TRUE,
                     split.group=T,
                     map.symbol = T,
                 is.signal=F,
                     kegg.native = F, 
                     match.data = T, 
                 node.sum='max.abs',
                     multi.state = T, 
                     bins=20,
                     same.layer = F,
                     pdf.size=c(7,7),
                     limit = list(gene=5, cpd=1))
```



Of course, you might want more control over the process. In that case, we can use another package. For example, we could use `clusterProfiler`. In this case, we need to manually select the genes that we want to run KEGG enrichment on.

```{r}
# turn res object into a list of genes
DE_genes <- res %>% data.frame() %>%
  filter(PValue < 0.01 & abs(logFC)>0.5
         ) %>%
  pull(ORF)
```


To run the enrichment, we can use the `enrichKEGG()` function from the `clusterProfiler` package. The argument "gene" for this funciton requires a vector of entrez gene id's. Right now we have a vector of gene IDs, but they are the ORF names instead of entrez IDs, so for most organisms we neeed to convert them.

Thankfully, the `clusterProfiler` package comes with the function `bitr` (Biological Id TRanslator) to translate geneIDs.

Note: the `sce` genome database is coded differently than most genome databases, so it requests the `ORF` instead of the `entrezID`, so we can directly use the DE_genes vector instead of the entrez ID. If you don't work with yeast, you'll probably need to use the entrez ID list. 

```{r}
# convert gene IDs to entrez IDs
entrez_ids <- bitr(DE_genes, fromType = "ORF", toType = "ENTREZID", OrgDb = org.Sc.sgd.db)

# Run our KEGG enrichment
kegg_results <- enrichKEGG(gene = DE_genes, #entrez_ids$ENTREZID, 
                           organism = 'sce' #options: https://www.genome.jp/kegg/catalog/org_list.html
                           )

# take a peak at the results
head(kegg_results)
data.frame(kegg_results) %>% View()


# Remove " - Saccharomyces cerevesiae" from each description entry
kegg_results@result$Description <- kegg_results@result$Description %>% print() %>% str_replace_all(., fixed(" - Saccharomyces cerevisiae"), "")
```
The first way we might want to visualize this plot is part of a KEGG pathway. We can open an html window with genes that are enriched highlighted, like this:
```{r}
browseKEGG(kegg_results, 'sce00500')
browseKEGG(kegg_results, 'sce04213')
```


clusterProfiler also let's us visualize these results
```{r fig.height=8, fig.width=8}
# Plot the KEGG pathway enrichment results
dotplot(kegg_results, showCategory = 10)

# cnetplot
cnetplot(kegg_results,  
         showCategory=10,
         node_label="category",
         color.params=list(foldChange = fold_change_geneList),
         cex.params=list(gene_label = 0.1,
                         category_label=0.5), 
         max.overlaps=100) +
  # change the color scale
  scale_colour_gradientn(colours = RColorBrewer::brewer.pal(n = 10, name = "RdYlBu"), limits=c(-6, 6)) +
  labs(colour="logFC")

# circulate cnetplot
cnetplot(kegg_results,  
         showCategory=6,
         circular=TRUE, # this changes the output graphing
         node_label="category",
         color.params=list(foldChange = fold_change_geneList),
         cex.params=list(gene_label = 0.5,
                         category_label=0.5), 
         max.overlaps=100) +
  # change the color scale
  scale_colour_gradientn(colours = RColorBrewer::brewer.pal(n = 10, name = "RdYlBu"), limits=c(-6, 6)) +
  labs(colour="logFC")

# heatplot
heatplot(kegg_results, showCategory=18, foldChange = fold_change_geneList)


```

emapplot
```{r}
x2 = enrichplot::pairwise_termsim(kegg_results)
emapplot(x2, showCategory = 30,cex_label_category = 0.4,)
```


## KEGG pathway gene set enrichment analysis

```{r}
# For gseKEGG, we need a gene list with na removed and sorted. let's do that now.
# omit any NA values 
kegg_gene_list = na.omit(fold_change_geneList)

# sort the list in decreasing order (required for clusterProfiler)
kegg_gene_list = sort(kegg_gene_list, decreasing = TRUE)

gse_kegg <- gseKEGG(geneList     = kegg_gene_list,
               organism     = 'sce',
               # minGSSize    = 120,
               pvalueCutoff = 0.05,
               verbose      = FALSE)
head(gse_kegg)

View(data.frame(gse_kegg))


# replace gse_kegg too long of names
gse_kegg@result$Description <- gse_kegg@result$Description %>% str_replace_all(., fixed(" - Saccharomyces cerevisiae"), "")
```
visualize those results
```{r fig.height=6}
dotplot(gse_kegg, showCategory = 3, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)


# Visualize the ranking of the genes by set
gseaplot(gse_kegg, geneSetID = 1, by = "preranked", title = gse_kegg$Description[1])
gseaplot(gse_kegg, geneSetID = 2, by = "preranked", title = gse_kegg$Description[2])
gseaplot(gse_kegg, geneSetID = 3, by = "preranked", title = gse_kegg$Description[3])
gseaplot(gse_kegg, geneSetID = 4, by = "preranked", title = gse_kegg$Description[4])
```

```{r fig.height=6, fig.width=8, dpi=300}
# take a closer look at the RNA polymerase genes
gseaplot(gse_kegg, geneSetID = 4, title = gse_kegg$Description[4])

# what if we want to look at multiple together?
enrichplot::gseaplot2(gse_kegg, geneSetID = 1:3, pvalue_table = TRUE,
          color = c("#E495A5", "#86B875", "#7DB0DD"), ES_geom = "dot")
```


# comparing Paralogs in common pathways

```{r}
# p_load(KEGGgraph)
toyKGML <- "~/Documents/GitHub/GenomicDataAnalysis_Fa23/exercises/sce00010.xml"
toyGraph <- parseKGML2Graph(toyKGML, genesOnly=FALSE)
toyGraph

nodes(toyGraph)

plot(toyGraph)

test <- ggkegg(#toyGraph,
       pid="sce00010",
            # convert_org = c("pathway","eco"),
            delete_zero_degree = TRUE,
            return_igraph = TRUE)

ttest <- ggraph(test, layout="stress") 
ttest$data$type |> unique()
#> [1] "map"      "compound" "gene"
ttest + geom_edge_diagonal(
  aes(color=subtype_name,
      filter=type!="maplink"))+
  geom_node_point(
  aes(filter= !type%in%c("map","compound")),
    # fill=gg$data[!gg$data$type%in%c("map","compound"),]$bgcolor,
    color="black",
    shape=21, size=4
  )+
  geom_node_point(
    aes(filter= !type%in%c("map","gene")),
    # fill=gg$data[!gg$data$type%in%c("map","gene"),]$bgcolor,
    color="black",
    shape=21, size=6
  )+
  geom_node_text(
    aes(label=name,
        filter=type=="gene"),
    repel=TRUE,
    bg.colour="white")+
  theme_void()
```

```{r}
p_install("ggkegg",try.bioconductor = TRUE)

p_install_gh("noriakis/ggkegg")

if(require(org.Sc.sgd.db)) {
  tnodes <- nodes(toprSub)
  tgeneids <- translateKEGGID2GeneID(tnodes)
  tgenesymbols <- sapply(mget(tgeneids, org.Hs.egSYMBOL, ifnotfound=NA), "[[")
  toprSubSymbol <- toprSub
  nodes(toprSubSymbol) <- tgenesymbols
  plot(toprSubSymbol, "neato",attrs=list(node=list(font=5, fillcolor="lightblue")))
}
                          


graph.par(list(nodes=list(col="darkgreen", lty="dotted", lwd=2, fontsize=6)))
labels <- edgeNames(toyGraph)
names(labels) <- labels



layToyGraph <- layoutGraph(toyGraph)


renderGraph(layToyGraph)                

graph <- ggkegg::pathway("hsa04110", use_cache=TRUE)
graph
```

```{r}
# started ggkegg here I think

pathway("ko01100",use_cache = TRUE) |>
  process_reaction() |>
  highlight_module(module("M00021")) |>
  highlight_module(module("M00338")) |>
  ggraph(x=x, y=y) +
  geom_node_point(size=1, aes(color=I(fgcolor),
                              filter=fgcolor!="none" & type!="line"))+
  geom_edge_link0(width=0.1, aes(color=I(fgcolor),
                                filter=type=="line"& fgcolor!="none"))+
  with_outer_glow(
    geom_edge_link0(width=1,
                   aes(color=I(fgcolor),
                       filter=(M00021 | M00338))),
    colour="red", expand=5
  )+
  with_outer_glow(
    geom_node_point(size=1.5,
                    aes(color=I(fgcolor),
                        filter=(M00021 | M00338))),
    colour="red", expand=5
  )+
  geom_node_text(size=2,
                 aes(x=x, y=y,
                     label=graphics_name,
                     filter=name=="path:ko00270"),
                 repel=TRUE, family="sans", bg.colour="white")+
  theme_void()
```

```{r}


pathway("sce00010",use_cache = TRUE) |>
  process_line() |>
  # highlight_module(module("M00021")) |>
  # highlight_module(module("M00338")) |>
  ggraph(x=x, y=y) +
  geom_node_point(size=1, aes(color=I(fgcolor),
                              filter=fgcolor!="none" & type!="line"))+
  geom_edge_link0(width=0.1, aes(color=I(fgcolor),
                                filter=type=="line"& fgcolor!="none"))+
  with_outer_glow(
    geom_edge_link0(width=1,
                   aes(color=I(fgcolor),
                       filter=(M00021 | M00338))),
    colour="red", expand=5
  )+
  with_outer_glow(
    geom_node_point(size=1.5,
                    aes(color=I(fgcolor),
                        filter=(M00021 | M00338))),
    colour="red", expand=5
  )+
  geom_node_text(size=2,
                 aes(x=x, y=y,
                     label=graphics_name,
                     filter=name=="path:ko00270"),
                 repel=TRUE, family="sans", bg.colour="white")+
  theme_void()



g <- pathway("sce00010")
pseudo_lfc <- sample(seq(0,3,0.1), length(V(g)), replace=TRUE)
names(pseudo_lfc) <- V(g)$name



ggkegg(
  "sce00010",
  convert_org = c("pathway", "sce", "ko"),
  convert_collapse = " ",
  numeric_attribute = pseudo_lfc
) +
  geom_edge_parallel2(
    aes(color = subtype_name),
    arrow = arrow(length = unit(1, 'mm')),
    start_cap = square(1, 'cm'),
    end_cap = square(1.5, 'cm')
  ) +
  geom_node_rect(aes(filter = .data$type == "group"),
                 fill = "transparent",
                 color = "red") +
  geom_node_rect(aes(fill = numeric_attribute,
                     filter = .data$type == "gene")) +
  geom_node_text(
    aes(label = converted_name,
        filter = .data$type == "gene"),
    size = 2.5,
    color = "black"
  ) +
  with_outer_glow(
    geom_node_text(
      aes(label = converted_name,
          filter = converted_name == "PCNA"),
      size = 2.5,
      color = "red"
    ),
    colour = "white",
    expand = 4
  ) +
  scale_edge_color_manual(values = viridis::plasma(11)) +
  scale_fill_viridis(name = "LFC") +
  theme_void()


kegg_results %>% ggkegg(convert_collapse = ":") +
  geom_node_point(size=1) +
  geom_edge_link0()




g <- pathway("ko00520",convert_org = c("pathway","sce"))
V(g)$color_one <- colorRampPalette(RColorBrewer::brewer.pal(5,"Set1"))(length(V(g)))
V(g)$color_two <- colorRampPalette(RColorBrewer::brewer.pal(5,"Set2"))(length(V(g)))

ggraph(g, x=x, y=y) +
  geom_node_rect(aes(xmin=xmin, xmax=x, fill=I(color_one)), alpha=0.5)+
  geom_node_rect(aes(xmin=x, xmax=xmax, fill=I(color_two)), alpha=0.5)+
  geom_node_text(size=2,
                 aes(x=x, y=y,
                     label=graphics_name#,
                     # filter=name=="path:ko00270"
                     ),
                 # repel=TRUE,
                 family="sans", bg.colour="white")+
  # geom_edge_parallel()+
  ggfx::with_outer_glow(geom_node_text(aes(label=name |> 
                       strsplit(":") |> 
                       sapply("[", 2) |>
                       strsplit(" ") |>
                       sapply("[", 1),
                     filter=type=="ortholog"),
                     size=2), colour="white", expand=1)



kegg_results |>
  ggkegg(convert_org = "sce", pathway_number = 2,group_rect_nudge=0) +
    geom_edge_link(
    aes(color=subtype_name),
    arrow = arrow(length = unit(1, 'mm')), 
    start_cap = square(1, 'cm'),
    end_cap = square(1.5, 'cm')) + 
    # geom_node_rect(aes(filter=.data$undefined & !.data$type=="ortholog"),
    #                fill="transparent", color="red")+
    geom_node_rect(aes(xmin=xmin*.9, xmax=x*1.1,filter=!.data$undefined &
                         .data$type=="gene"), fill="white", color="black")+
    geom_node_text(aes(label=graphics_name,
                       filter=.data$type == "gene"),
                   size=1.5,
                   color="black",family="serif")+
    with_outer_glow(geom_node_text(aes(label=converted_name,
                                       filter=.data$enrich_attribute),
                                   size=2.5, color="red"),
                    colour="white",
                    expand=4)+
    theme_void()




g <- pathway("sce00010",group_rect_nudge=0) |> mutate(conv=convert_id("sce")) ## Obtain pathway graph (don't show grouping rect)

V(g)$color_one <- colorRampPalette(RColorBrewer::brewer.pal(5,"Set1"))(length(V(g)))
V(g)$color_two <- colorRampPalette(RColorBrewer::brewer.pal(5,"Set2"))(length(V(g)))

V(g)

gg <- ggraph(g, layout="manual", x=x, y=y)+
  geom_node_point(aes(filter=type=="compound"), color="blue", size=2)+
  # geom_node_rect(fill="red",aes(filter=type=="ortholog"))+
  geom_node_rect(aes(filter=type=="ortholog",xmin=xmin, xmax=x, fill=I(color_one)), alpha=0.5)+
  geom_node_rect(aes(filter=type=="ortholog",xmin=x, xmax=xmax, fill=I(color_two)), alpha=0.5)+
  overlay_raw_map("sce00010")+
  geom_node_text(
    aes(label=graphics_name,
        filter=type=="gene"),
    # repel=TRUE,
    bg.colour="white")+
  theme_void()
gg

gg$labels


g <- delete_vertex_attr(g, "x")
g <- delete_vertex_attr(g, "y")
ggraph(g, layout = "nicely") +
  geom_node_point(aes(filter=type=="gene" | type=="group"), color="black") +
  geom_edge_link(aes(color=subtype_name),end_cap=circle(2,"mm"),
                 start_cap=circle(2,"mm"),
                     label_dodge = unit(2,"mm"))+
  geom_node_text(aes(label=conv), repel=TRUE, bg.colour="white")+
  theme_void()


```

```{r}
# try TCA instead of glycolysis, see if any paralogs better labelled
g <- pathway("sce00020",group_rect_nudge=0) |> mutate(conv=convert_id("sce")) ## Obtain pathway graph (don't show grouping rect)

V(g)$color_one <- colorRampPalette(RColorBrewer::brewer.pal(5,"Set1"))(length(V(g)))
V(g)$color_two <- colorRampPalette(RColorBrewer::brewer.pal(5,"Set2"))(length(V(g)))

V(g)

gg <- ggraph(g, layout="manual", x=x, y=y)+
  geom_node_point(aes(filter=type=="compound"), color="blue", size=2)+
  # geom_node_rect(fill="red",aes(filter=type=="ortholog"))+
  geom_node_rect(aes(filter=type=="ortholog",xmin=xmin, xmax=x, fill=I(color_one)), alpha=0.5)+
  geom_node_rect(aes(filter=type=="ortholog",xmin=x, xmax=xmax, fill=I(color_two)), alpha=0.5)+
  # overlay_raw_map("sce00020")+
  geom_node_text(
    aes(label=graphics_name,
        filter=type=="gene"),
    size=2,
    # repel=TRUE,
    bg.colour="white")+
  theme_void()
gg



# let me try again, splitting graphics_name
graph <- pathway("sce00500")  |> mutate(showname=strsplit(name, " ") |> str_remove_all("sce:")) #|>
  
graph <- graph |> mutate(showname = gsub('c\\(|\\)|"|"|,', '', showname)) |>
  mutate(showname=strsplit(showname, " ") |>
                    vapply("[", 1, FUN.VALUE="a"))


ggraph(graph, layout="manual", x=x, y=y)+
    geom_edge_parallel(aes(linetype=subtype_name),
        arrow=arrow(length=unit(1,"mm"), type="closed"),
        end_cap=circle(1,"cm"),
        start_cap=circle(1,"cm"))+
    geom_node_rect(aes(fill=I(bgcolor),
                      filter=type == "ortholog"),
                  color="black")+
    geom_node_text(aes(label=showname,
                      filter=type == "ortholog"),
                  size=2)+
    theme_void()
```


# making a custom visualization
```{r}


graph <- pathway("sce00500")  |> mutate(showname=strsplit(name, " ") |> str_remove_all("sce:")) #|>
  
graph <- graph |> mutate(showname = gsub('c\\(|\\)|"|"|,', '', showname)) |>
  mutate(showname=strsplit(showname, " ") |>
                    vapply("[", 1, FUN.VALUE="a"))

data <- graph |>   mutate(
    test_orthologs=convert_id("ko", first_arg_sep= FALSE))
    


# i saw here that instead of gene, we can use ortholog as our filter type here.
ggraph(data, layout="manual", x=x, y=y)+
    geom_edge_parallel(aes(linetype=subtype_name),
        arrow=arrow(length=unit(1,"mm"), type="closed"),
        end_cap=circle(1,"cm"),
        start_cap=circle(1,"cm"))+
    overlay_raw_map("sce00500")+
    geom_node_rect(aes(fill=I(bgcolor),
                      filter=type == "ortholog"),
                  color="black")+
    geom_node_text(aes(label=test_orthologs,
                      filter=type == "ortholog"),
                  size=2)+
    theme_void()
  
  
```



```{r}
tst <- ggkegg(
  pid= "sce00010",
  return_igraph = TRUE,
  # convert_org = c("sce", "ko", "pathways"),
  convert_first = FALSE,
  # delete_zero_degree = TRUE,
  # delete_undefined = TRUE
)

tst

kegg_results %>%
  ggkegg(#toyGraph,
       # pid="sce00010",
      convert_first = FALSE,
      convert_collapse = "\n",
    pathway_number = 2,
            convert_org = c("sce"),
            delete_zero_degree = TRUE,
            return_igraph = FALSE)+
  # geom_node_rect(aes(.data$type=="gene",ymin=y-20, ymax=y+2), position = , fill="white", color="black") +
  geom_node_text(aes(label=converted_name,
                       filter=.data$type == "gene"),
                   size=2.5,
                   color="black",family="serif") 
      
test <- kegg_results %>%
  ggkegg(
       # pid="sce00010",
      convert_first = FALSE,
      convert_collapse = "\n",
    pathway_number = 20,
            convert_org = c("sce"),
            delete_zero_degree = TRUE,
            return_igraph = FALSE)


test$data$n_genes <- str_count(test$data$name, '\\w+')/2
# 
# test +
#   # geom_node_rect(aes(filter=.data$type=="gene"),fill="white", color="black") +
#   # geom_node_rect(aes(filter=.data$type=="gene",ymin=y-4*(y-ymin), ymax=ymin, y=y-2.5*(y-ymin)), fill="white", color="black") +
#   # geom_rect(data)
#   geom_rect(data = filter(test$data, type=="gene" & n_genes == 1), mapping = aes(xmin=xmin, ymin=ymin, xmax=xmax, ymax=ymax)) +
#   geom_rect(data = filter(test$data, type=="gene" & n_genes == 1),mapping = aes(xmin=xmin, ymin=ymin, xmax=xmax, ymax=ymax)) +
#   geom_node_text(aes(label=converted_name,
#                        filter=.data$type == "gene"),
#                    size=2.5,
#                    color="black",family="serif")



res

tmp <- test$data


# where it came from: "https://rest.kegg.jp/list/reaction/"
KEGG_reactions <- KEGGREST::keggList("reaction") |> as.data.frame() |>
  rownames_to_column("reaction") |>
  mutate(reaction_description = str_split_i(`KEGGREST::keggList("reaction")`, ";", 1)) |>
  select(reaction, reaction_description) |>
  mutate(reaction=paste0('rn:', reaction))

tmp |>
  filter(type == "gene") |>
  mutate(showname=strsplit(name, " ") |> str_remove_all("sce:")) |> 
  mutate(showname = gsub('c\\(|\\)|"|"|,', '', showname)) |>
  separate_rows(showname, sep = " ") |>
  left_join(rownames_to_column(res$table), by=c("showname" = "rowname")) |>
  left_join(KEGG_reactions, by="reaction") |>
  mutate(gene_name= AnnotationDbi::select(org.Sc.sgd.db,keys=showname,columns="GENENAME")$GENENAME) |>
  mutate(gene_name = coalesce(gene_name,showname)) |>
  ggplot(aes(x=x,y=y)) +
  overlay_raw_map("sce00020")+
  ggrepel::geom_label_repel(aes(label = gene_name, fill = logFC), box.padding = 0.01,direction = "y",
                            size=2,max.overlaps = 15#,position=position_jitter()
                            ) +
  # geom_text(aes(label = showname),position = position_stack(vjust = -0.01),)
  theme_void() +
  # facet_wrap(~reaction_description,labeller = labeller(reaction_description = label_wrap_gen(10)),
             # scales = "free") +
  scale_fill_gradientn(colours = RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")#, limits=c(-6, 6)
                       )
```
Let's look at the paralogs in glycolysis KEGG pathway
```{r}
# 24th path in kegg_results
glycolysis_KEGG <- kegg_results %>%
  ggkegg(
       # pid="sce00010",
      convert_first = FALSE,
      convert_collapse = "\n",
    pathway_number = 24, # the 24th entry in kegg_results is glycolysis, change this to change the pathway.
            convert_org = c("sce"),
            delete_zero_degree = TRUE,
            return_igraph = FALSE)


glycolysis_KEGG$data |>
  filter(type == "gene") |>
  mutate(showname=strsplit(name, " ") |> str_remove_all("sce:")) |> 
  mutate(showname = gsub('c\\(|\\)|"|"|,', '', showname)) |>
  separate_rows(showname, sep = " ") |>
  left_join(rownames_to_column(res$table), by=c("showname" = "rowname")) |>
  left_join(KEGG_reactions, by="reaction") |>
  mutate(gene_name= AnnotationDbi::select(org.Sc.sgd.db,keys=showname,columns="GENENAME")$GENENAME) |>
  mutate(gene_name = coalesce(gene_name,showname)) |>
  ggplot(aes(x=x,y=y)) +
  overlay_raw_map("sce00010")+
  ggrepel::geom_label_repel(aes(label = gene_name, fill = logFC), box.padding = 0.01,label.padding=0.05,direction = "y",
                            size=3,max.overlaps = 15,color = "black",label.r=0.01
                # bg.color = "white",#,position=position_jitter()
                            ) +
  # geom_text(aes(label = showname),position = position_stack(vjust = -0.01),)
  theme_void() +
  # facet_wrap(~reaction_description,labeller = labeller(reaction_description = label_wrap_gen(10)),
             # scales = "free") +
  scale_fill_gradientn(colours = RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")#, limits=c(-6, 6)
                       )
```
Try it with the res_all object instead, so we can compare logFC across contrasts.
```{r}
glycolysis_KEGG$data |>
  filter(type == "gene") |>
  mutate(showname=strsplit(name, " ") |> str_remove_all("sce:")) |> 
  mutate(showname = gsub('c\\(|\\)|"|"|,', '', showname)) |>
  separate_rows(showname, sep = " ") |>
  left_join(rownames_to_column(res_all$table), by=c("showname" = "rowname")) |>
  left_join(KEGG_reactions, by="reaction") |>
  mutate(gene_name= AnnotationDbi::select(org.Sc.sgd.db,keys=showname,columns="GENENAME")$GENENAME) |>
  mutate(gene_name = coalesce(gene_name,showname)) |> #View()
  ggplot(aes(x=x,y=y)) +
  overlay_raw_map("sce00010")+
  ggrepel::geom_label_repel(aes(label = gene_name, fill = logFC.EtOHvsMOCK.WT, color=logFC.EtOHvsMOCK.MSN24dd), box.padding = 0.05,label.padding=0.05,direction = "y",label.size = 1.8,
                            size=3,max.overlaps = 100,label.r=0.002,seed=123
                # bg.color = "white",#,position=position_jitter()
                            ) +
  ggrepel::geom_label_repel(aes(label = gene_name, fill = logFC.EtOHvsMOCK.WT), color="black", box.padding = 0.05,label.padding=0.05,direction = "y",label.size = NA,
                            size=3,max.overlaps = 100,label.r=0.02,seed=123
                # bg.color = "white",#,position=position_jitter()
                            ) +
  # geom_text(aes(label = showname),position = position_stack(vjust = -0.01),)
  theme_void() +
  # facet_wrap(~reaction_description,labeller = labeller(reaction_description = label_wrap_gen(10)),
             # scales = "free") +
  scale_fill_gradientn(colours = RColorBrewer::brewer.pal(n = 10, name = "RdYlBu"), breaks=c(-6, 6)) +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 10, name = "RdYlBu"), breaks=c(-6, 6)
                       ) +
  ggtitle("EtOH response of WT and Msn2/4∆ (border: Msn2/4∆, fill: WT)")
```

Side by side of the WT EtOH response and the Msn2/4∆∆ EtOH response
```{r}
set.seed=123
pathway_to_graph <- "sce00010" #glycolysis/gluconeogenesis
# pathway_to_graph <- "sce00020" # TCA cycle
# pathway_to_graph <- "sce04213" #longevity
# pathway_to_graph <- "sce00500" # starch & sucrose metabolism
# pathway_to_graph <- "sce00620" # pyruvate metabolism

pathway_number <- kegg_results %>% 
  data.frame() |> 
  mutate(row_number = row_number()) |> 
  filter(ID == pathway_to_graph) %>%
  pull(row_number)

KEGG_data <- kegg_results %>%
  ggkegg(
    # pid="sce00010",
    convert_first = FALSE,
    convert_collapse = "\n",
    pathway_number = pathway_number, # changing this to change the pathway.
    convert_org = c("sce"),
    delete_zero_degree = TRUE,
    return_igraph = FALSE
  )

WT.graph <- KEGG_data$data |>
  filter(type == "gene") |>
  mutate(showname=strsplit(name, " ") |> str_remove_all("sce:")) |> 
  mutate(showname = gsub('c\\(|\\)|"|"|,', '', showname)) |>
  separate_rows(showname, sep = " ") |>
  left_join(rownames_to_column(res_all$table), by=c("showname" = "rowname")) |>
  left_join(KEGG_reactions, by="reaction") |>
  mutate(gene_name= AnnotationDbi::select(org.Sc.sgd.db,keys=showname,columns="GENENAME")$GENENAME) |>
  mutate(gene_name = coalesce(gene_name,showname)) |> #View()
  ggplot(aes(x=x,y=y)) +
  overlay_raw_map(pathway_to_graph)+
  ggrepel::geom_label_repel(aes(label = gene_name, fill = logFC.EtOHvsMOCK.WT), box.padding = 0.05,label.padding=0.05,direction = "y",#label.size = 1.8,
                            size=2,max.overlaps = 100,label.r=0.002,seed=123
                # bg.color = "white",#,position=position_jitter()
                            ) +
  # ggrepel::geom_label_repel(aes(label = gene_name, fill = logFC.EtOHvsMOCK.WT), color="black", box.padding = 0.05,label.padding=0.05,direction = "y",label.size = NA,
  #                           size=3,max.overlaps = 100,label.r=0.02,seed=123
                # bg.color = "white",#,position=position_jitter()
                            # ) +
  # geom_text(aes(label = showname),position = position_stack(vjust = -0.01),)
  theme_void() +
  # facet_wrap(~reaction_description,labeller = labeller(reaction_description = label_wrap_gen(10)),
             # scales = "free") +
  scale_fill_gradientn(colours = rev(RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")), limits=c(-10.5,10.5)
                       ) +
  # 
  # scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 10, name = "RdYlBu"), breaks=c(-6, 6)
                       # ) +
  # theme(plot.margin=margin(l=-0.8,unit="cm"), legend.margin = margin(l=-0.8,unit="cm")) +
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm")#,legend.position="none"
        ) +
  labs(fill="logFC", title="      WT Response to Ethanol")

msn24.graph <- KEGG_data$data |>
  filter(type == "gene") |>
  mutate(showname=strsplit(name, " ") |> str_remove_all("sce:")) |> 
  mutate(showname = gsub('c\\(|\\)|"|"|,', '', showname)) |>
  separate_rows(showname, sep = " ") |>
  left_join(rownames_to_column(res_all$table), by=c("showname" = "rowname")) |>
  left_join(KEGG_reactions, by="reaction") |>
  mutate(gene_name= AnnotationDbi::select(org.Sc.sgd.db,keys=showname,columns="GENENAME")$GENENAME) |>
  mutate(gene_name = coalesce(gene_name,showname)) |> #View()
  ggplot(aes(x=x,y=y)) +
  overlay_raw_map(pathway_to_graph)+
  ggrepel::geom_label_repel(aes(label = gene_name, fill = logFC.EtOHvsMOCK.MSN24dd), box.padding = 0.05,label.padding=0.05,direction = "y",#label.size = 1.8,
                            size=2,max.overlaps = 100,label.r=0.002,seed=123
                # bg.color = "white",#,position=position_jitter()
                            ) +
  # ggrepel::geom_label_repel(aes(label = gene_name, fill = logFC.EtOHvsMOCK.WT), color="black", box.padding = 0.05,label.padding=0.05,direction = "y",label.size = NA,
  #                           size=3,max.overlaps = 100,label.r=0.02,seed=123
                # bg.color = "white",#,position=position_jitter()
                            # ) +
  # geom_text(aes(label = showname),position = position_stack(vjust = -0.01),)
  theme_void() +
  guides(fill = FALSE) +
  # facet_wrap(~reaction_description,labeller = labeller(reaction_description = label_wrap_gen(10)),
             # scales = "free") +
  scale_fill_gradientn(colours = rev(RColorBrewer::brewer.pal(n = 10, name = "RdYlBu")), limits=c(-10.5,10.5)
                       ) +
  # scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 10, name = "RdYlBu"), breaks=c(-6, 6)
                       # )
  
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  labs(fill="logFC", title="      Msn2/4∆∆ Response to Ethanol")




WT.graph + msn24.graph +
  patchwork::plot_annotation(tag_levels = 'A')# & 
  # theme(plot.tag = element_text(size = 8))

```


Be sure to knit this file into a pdf or html file once you're finished.

System information for reproducibility:
```{r}
pander::pander(sessionInfo())
```
