---
title: "Differential Expression Analysis with EdgeR"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output: html_document
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
set.seed("1492")
```

# Getting Things Setup

As usual, make sure we have the right packages for this exercise

```{r ready-packages}
if (!require("pacman")) install.packages("pacman"); library(pacman)

# let's load all of the files we were using and want to have again today
p_load("tidyverse", "knitr", "readr",
       "pander", "BiocManager", 
       "dplyr", "stringr", 
       "purrr", # for working with lists (beautify column names)
       "reactable") # for pretty tables.

# We also need these Bioconductor packages today.
p_load("edgeR", "AnnotationDbi", "org.Sc.sgd.db", "pathview",
       "clusterProfiler")# for dotplot
```


# Description


In this class exercise, we will explore the use of KEGG pathways in genomic data analysis. KEGG is a valuable resource for understanding biological pathways and functions associated with genomic data.

# Learning outcomes

At the end of this exercise, you should be able to:

-   Understand the basics of KEGG pathways.
-   Learn how to retrieve and analyze pathway information using R.
-   Apply KEGG analysis to genomic data.


```{r load-libraries}
library(edgeR)
library(org.Sc.sgd.db)
# for ease of use, set max number of digits after decimal
options(digits=3)
```

# Loading in the edgeR DE gene file output.


```{r save-res}
# Choose topTags destination
dir_output_edgeR <-
  path.expand("~/Desktop/Genomic_Data_Analysis/Analysis/edgeR/")

# for subsequent analysis, let's save the res object as an R data object.
res <- read_rds(file = paste0(dir_output_edgeR, "yeast_res_edgeR.Rds"))
```



# KEGG

First, let's do a simple KEGG enrichment with the limma package, using the function `kegga()`. This approach is nice, because we can directly load in the `res` object we created in the `edgeR` workflow, set our FDR cutoff, and the process happens automatically.

```{r}
# alias2Symbol()

# go <- goana(res, species.KEGG="sce")
# topGO(go, n=10)

# from limma, there is the 
kegg <- kegga(res, species.KEGG="sce", FDR=0.01,
      plot=T)

topKEGG(kegg, sort="up")


gene_data_logFC <- res$table %>% dplyr::select(logFC)


if (!requireNamespace("pathview", quietly = TRUE))
    BiocManager::install("pathview")

library("pathview")
test <- pathview(gene.data  = gene_data_logFC,
                     # pathway.id = "sce01100", #metabolic pathways
                     pathway.id = "sce00010", #glycolysis/gluconeogenesis
                     # pathway.id = "sce03050", # proteasome cycle
                     # pathway.id = "sce00020", # TCA cycle
                     # pathway.id = "sce00500", # starch & sucrose metabolism (trehalose)
                     # pathway.id = "sce00030", #PPP
                     species    = "sce",
                     gene.idtype="orf",
                     expand.node = TRUE,
                     map.symbol = T,
                     kegg.native = F, 
                     match.data = T, 
                     multi.state = T, 
                     bins=20,
                     same.layer = F,
                     pdf.size=c(7,7),
                     limit = list(gene=5, cpd=1))
```



Of course, you might want more control over the process. In that case, we can use another package. For example, we could use `clusterProfiler`. In this case, we need to manually select the genes that we want to run KEGG enrichment on.

```{r}
# turn res object into a list of genes
DE_genes <- res %>% data.frame() %>%
  filter(PValue < 0.01 #& abs(logFC)>1
         ) %>%
  pull(ORF)
```


To run the enrichment, we can use the `enrichKEGG()` function from the `clusterProfiler` package. The argument "gene" for this funciton requires a vector of entrez gene id's. Right now we have a vector of gene IDs, but they are the ORF names instead of entrez IDs, so for most organisms we neeed to convert them.

Thankfully, the `clusterProfiler` package comes with the function `bitr` (Biological Id TRanslator) to translate geneIDs.

Note: the `sce` genome database is coded differently than most genome databases, so it requests the `ORF` instead of the `entrezID`, so we can directly use the DE_genes vector instead of the entrez ID. If you don't work with yeast, you'll probably need to use the entrez ID list. 

```{r}
# convert gene IDs to entrez IDs
entrez_ids <- bitr(DE_genes, fromType = "ORF", toType = "ENTREZID", OrgDb = org.Sc.sgd.db)

# Run our KEGG enrichment
kegg_results <- enrichKEGG(gene = DE_genes, #entrez_ids$ENTREZID, 
                           organism = 'sce' #options: https://www.genome.jp/kegg/catalog/org_list.html
                           )

# take a peak at the results
head(kegg_results)
data.frame(kegg_results) %>% View()


# Remove " - Saccharomyces cerevesiae" from each description entry
kegg_results@result$Description <- kegg_results@result$Description %>% print() %>% str_replace_all(., fixed(" - Saccharomyces cerevisiae"), "")
```

clusterProfiler also let's us visualize these results
```{r}
# Plot the KEGG pathway enrichment results
dotplot(kegg_results, showCategory = 10)

# cnetplot
cnetplot(kegg_results, categorySize="pvalue", foldChange=res %>% data.frame() %>%
  filter(PValue < 0.01 #& abs(logFC)>1
         ) %>% pull(logFC))
```

Be sure to knit this file into a pdf or html file once you're finished.

System information for reproducibility:
```{r}
pander::pander(sessionInfo())
```
