---
title: "Read Mapping"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output: html_document
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = "~/Desktop")
set.seed("1492")
```


As usual, make sure we have the right packages for this exercise

```{r ready-packages, results="hide",message=FALSE}
list.of.packages <- c("tidyverse", "here", "knitr", "pander", "BiocManager")

new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages,repos = "http://cran.us.r-project.org")
lapply(list.of.packages, require, character.only = TRUE)

```

We also need the bioconductor package "Rsubread" for today's activity.

```{r install-BiocPackages, results="hide", message=FALSE, warning=FALSE}
# make sure we have the CRAN package installed for getting bioconductor packages
if (!requireNamespace("Rsubread", quietly = TRUE))
    BiocManager::install("Rsubread")
```

Previously, we filtered and trimmed our raw fastq files. They should be in your ~/Desktop/filtered folder unless you chose a different place to store them.
```{r}

trimmed.fq_file_dir <- "~/Desktop/filtered"

trimmed.fastq.files <- list.files(path = trimmed.fq_file_dir, 
                                  pattern = ".fastq.gz$", 
                                  full.names = TRUE)
trimmed.fastq.files
```

You should see the full paths to all 8 trimmed fastq files that we will be mapping to the reference genome today.

# Alignment

Read sequences are stored in compressed (gzipped) FASTQ files. Before the differential expression analysis can proceed, these reads must be aligned to the yeast genome and counted into annotated genes. This can be achieved with functions in the Rsubread package.

## Retrieve the genome

We will use a bash code chunk to download the latest genome
```{bash fetchGenome}
# Define the destination file path
# Be sure to change this file path to the path you want your data to go
REF_DIR="/Users/$USER/Desktop/reference_genome"

# make that directory if it doesn't already
mkdir $REF_DIR

# Define the URL of reference genome
# (latest from ensembl)
url="ftp://ftp.ensembl.org/pub/release-110/fasta/saccharomyces_cerevisiae/dna/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa.gz"


# Check if the file already exists at the destination location
if [ ! -f "$REF_DIR/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa.gz" ]; then
    echo "Reference genome not found, downloading..."
    # If the file does not exist, download it using curl
    curl -o "$REF_DIR/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa.gz" "$url"
    echo "Downloading finished"
else
    echo "File already exists at $REF_DIR Skipping download."
fi
```


## Build the index

The first step in performing the alignment is to build an index. In order to build an index you need to have the fasta file (.fa), which can be downloaded from the UCSC genome browser.  This may take several minutes to run. Building the full index using the whole genome usually takes about 30 minutes to an hr on a server for larger Eukaryotic genomes. Because yeast has a relatively small genome size, we are able to build the full index in class.

```{r}
library(Rsubread)

# Set path of the reference fasta file
reference_genome = "/Users/clstacy/Desktop/reference_genome/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa.gz"

index_reference_genome = "/Users/clstacy/Desktop/reference_genome/sacCer3"

# build the index
buildindex(basename=index_reference_genome, reference=reference_genome)
```

We can see the arguments available with the align function from the Rsubread package
```{r argsAlign}
args(align)
```


```{r alignReads}
# set where we would like the file to be output
##NOTE: I don't think we need this. but if it doesn't put them in the same folder where the trimmed fastq files, then maybe setwd in this chunk.
# setwd("~/Desktop")

# run the alignment on all of the trimmed.fastq.files
align(index=index_reference_genome, 
      readfile1=trimmed.fastq.files,
      type = "rna",
      input_format = "gzFASTQ",
      output_format = "BAM",
      unique = FALSE,
      nBestLocations = 1
      )
```

The output of the alignment are bam corresponding to each fastq file. 

We can get a summary of the proportion of reads that mapped to the reference genome using the propmapped function.

```{r summaryOutput}
# create an object in R listing
bam.files <- list.files(path = "~/Desktop/filtered", pattern = ".BAM$", full.names = TRUE)
bam.files

# find the proportion of reads that mapped for each sample
props <- propmapped(files=bam.files)

props
```




Question 1: Try aligning the fastq files allowing multi-mapping reads (set unique = FALSE), and allowing for up to 6 “best” locations to be reported (nBestLocations = 6). Specify the output file names (bam.files.multi) by substituting “.fastq.gz” with “.multi.bam” so we don’t overwrite our unique alignment bam files.

Question 2: Look at the proportion of reads mapped and see if we get any more reads mapping by specifying a less stringent criteria.









Be sure to knit this file into a pdf or html file once you're finished.

System information for reproducibility:

```{r}
pander::pander(sessionInfo())
```

