---
title: "Working with Sequences: Raw Data & Quality Control"
author: "Carson Stacy & Jeffrey Lewis"
date: "Fall 2023"
output: html_document
---

last updated: `r Sys.Date()`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
set.seed("832")
```

# Getting Things Setup

As usual, make sure we have the right packages for this exercise

```{r ready-packages}
if (!require("pacman")) install.packages("pacman"); library(pacman)

# let's load all of the files we were using and want to have again today
p_load("tidyverse", "knitr", "readr",
       "pander", "BiocManager", 
       "dplyr", "stringr")

# We also need the bioconductor packages "ShortRead" and "rfastp" for today's activity.
p_load("Rfastp", "Shortread")
```

# Description

This activity is intended to familiarize you with raw bioinformatic sequence files. Specifically, we'll be working with short read sequencing data generated from an Illumina platform.

# Learning outcomes

At the end of this exercise, you should be able to:

-   Load and read into R a raw gzipped fastq file.
-   Inspect sequence quality and evaluate results.
-   Perform quality control on raw data and save processed output.

Note that instead of {r}, the below chunk uses {bash}, meaning this isn't r code but bash code (the language used in the terminal). The -nc flag ensures the files are only downloaded if they don't already exist where you are downloading them.

This may take awhile. The below script is a bash command that downloads these files to your computer

```{bash download-fq, message=FALSE}
# Be sure to change this file path to the path you want your data to go
DATA_DIR="/Users/$USER/Desktop/Genomic_Data_Analysis/Raw_Data/"

mkdir -p $DATA_DIR
# WARNING: curl doesn't work with relative paths
curl -L -C - -o $DATA_DIR/subset_phenol_heat_rep1.fastq.gz https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/heat_shock/subset_phenol_heat_rep1.fastq.gz?raw=TRUE
curl -L -C - -o $DATA_DIR/subset_phenol_heat_rep2.fastq.gz https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/heat_shock/subset_phenol_heat_rep2.fastq.gz?raw=TRUE
curl -L -C - -o $DATA_DIR/subset_phenol_heat_rep3.fastq.gz https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/heat_shock/subset_phenol_heat_rep3.fastq.gz?raw=TRUE
curl -L -C - -o $DATA_DIR/subset_phenol_heat_rep4.fastq.gz https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/heat_shock/subset_phenol_heat_rep4.fastq.gz?raw=TRUE

curl -L -C - -o $DATA_DIR/subset_phenol_unstressed_rep1.fastq.gz https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/heat_shock/subset_phenol_unstressed_rep1.fastq.gz?raw=TRUE
curl -L -C - -o $DATA_DIR/subset_phenol_unstressed_rep2.fastq.gz https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/heat_shock/subset_phenol_unstressed_rep2.fastq.gz?raw=TRUE
curl -L -C - -o $DATA_DIR/subset_phenol_unstressed_rep3.fastq.gz https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/heat_shock/subset_phenol_unstressed_rep3.fastq.gz?raw=TRUE
curl -L -C - -o $DATA_DIR/subset_phenol_unstressed_rep4.fastq.gz https://github.com/clstacy/GenomicDataAnalysis_Fa23/raw/main/data/heat_shock/subset_phenol_unstressed_rep4.fastq.gz?raw=TRUE

# where these files came from: A current project in our lab.

# Let's see what one of these files contains:
gzcat $DATA_DIR/subset_phenol_unstressed_rep4.fastq.gz | head -n8
```

We have the data downloaded onto our system now, so let's first take a look at some of these files ourselves

The R package ShortRead allows us to look at and process raw fastq files. It has many more features than we will use today.

## Let's take a look at a fastq file

```{r}
# change this directory here to where you have the file saved
fastq_phenol_heat_rep4 <- readFastq("~/Desktop/subset_phenol_heat_rep4.fastq.gz")
# fastq_phenol_heat_rep4 <- readFastq(path_fastq_phenol_heat_rep4)

# file too big? swap readFastq() for:
fastq_phenol_unstressed_rep1 <- yield(FastqSampler("~/Desktop/subset_phenol_heat_rep4.fastq.gz", n=10000))

```

A few quick ways to examine the fastq data object

```{r}
# Typing the name of the object gives us a simple summary
fastq_phenol_heat_rep4

# the length() function gives us the total number of reads
length(fastq_phenol_heat_rep4)

# We can use the width() function to find the size of each read/sequence in fastq
width(fastq_phenol_heat_rep4) %>% head() # add head() pipe to only print first 10


#sread() - Retrieve sequence of reads.
sread(fastq_phenol_heat_rep4)

#quality() - Retrieve quality of reads as ASCII scores.
quality(fastq_phenol_heat_rep4)

#id() - Retrieve IDs of reads
id(fastq_phenol_heat_rep4)
```

The output of sread() is a DNAStringSet object, so we can use all of the commands from the biostrings library on the output

```{r}
# first, let's save the output of sread as an object
sequence_of_reads <- sread(fastq_phenol_heat_rep4)

# Now, let's use the biostrings function alphabetFrequency to see
# the occurrence of nucleotide bases in reads.
alph_freq <- alphabetFrequency(sequence_of_reads)

# subset just the first two reads
alph_freq[1:2,]
```

We see most of the nucleotides are assigned to A, C, G, or T, with one base in each read an N.

A fundamental difference between fasta and fastq files is the Quality scores containined in fastQ.

Quality scores are stored as ASCII characters representing -log10 probability of base being wrong (Larger scores would be associated to more confident base calls).

A comprehensive description of phred quality can be found on the wiki page for FastQ.

To see the fastq encodings, we can run:

```{r}
encoding(quality(fastq_phenol_heat_rep4))
```

The ShortRead package has many functions available to allow us to collect useful metrics from our ShortRead object.

One very useful function is the alphabetByCycle() function which provides a quick method to summarise base occurrence of cycles.

Here we apply alphabetByCycle() function to the sequence information and show the occurrence of main 4 bases over first 15 cycles.

```{r}
alph_by_cycle <- alphabetByCycle(sequence_of_reads)
alph_by_cycle[1:4,1:15]
```

We can use the table function to identify the number of times a sequence appears in our FastQ file's sequence reads.

```{r}
readOccurence <- table(sequence_of_reads)

# see the top 3 sequences that appear the highest number of times
sort(readOccurence,decreasing = TRUE)[1:3]
```

We can identify duplicated reads (potentially arising from PCR over amplification) by using the srduplicated() function and the ShortReadQ object.

This returns a logical vector identifying which reads' sequences are duplicates (occur more than once in file). Note that the first time a sequence appears in file is not a duplicate but the second, third, fourth times etc are.

```{r}
duplicates <- srduplicated(fastq_phenol_heat_rep4)
duplicates[1:3]

# we can use table() to get a quick summary of the seq duplication rate
table(duplicates)
```

The ShortRead package also contains a function to generate a simple quality control report.

The qa() function accepts a FastQ file and returns a FastqQA object.

```{r}
qa_phenol_heat_rep4 <- qa("~/Desktop/subset_phenol_heat_rep4.fastq.gz")
qa_phenol_heat_rep4
```

We can then use the report() function to generate a simple report.

```{r}
myReport_phenol_heat_rep4 <- report(qa_phenol_heat_rep4)
myReport_phenol_heat_rep4
```

Finally we can review the report in a browser or use the browseURL function to open it in a browser from R.

```{r viewReport, eval=F}
browseURL(myReport_phenol_heat_rep4)
```

# Trimming

When we observe low quality at the end of reads we may wish to remove the low quality bases for later alignment to the genome. The trimTails() function trims reads from the 3', removing bases which fall below a desired quality. The trimTails() function accepts arguments specifying the ShortReadQ object, the minimum number of successive bases required to be below quality cut-off for trimming and the actual cut-off score.

```{r}
trimmed_fastq_phenol_heat_rep4 <- trimTails(fastq_phenol_heat_rep4, # ShortReadQ object to trim
                          k=10, # integer number of failing letters to trigger trim
                          a="5") # character giving letter at or below to "fail"
trimmed_fastq_phenol_heat_rep4
```

Now we have trimmed our FastQ reads, we can export these reads for further analysis using the writeFastq() function

```{r eval=FALSE}
writeFastq(trimmed_fastq_phenol_heat_rep4,
           "~/Desktop/phenol_heat_rep4_trimmed.fastq.gz") #path to save file
```

## Is there any way we can automate some of this?

There are several utility programs that will provide you with QC and trim your data for you, with less input from you. We like fastp as it does some basic QC and trims your fastq files, and it does it very quickly. To make this available in R, it has been made available in the Bioconductor package Rfastp.

By default, fastp will make a html report to summarize your result. But the Rfastp wrapper allows you to look at some of them in R.

```{r}
# if we wanted to just run a single file, we would do so like this:

rfastp_report <- rfastp(read1 = "/Users/clstacy/Desktop/subset_phenol_heat_rep1.fastq.gz",
                        outputFastq ="/Users/clstacy/Desktop/rfastp_trimmed_phenol_heat_rep1.fastq.gz")

df_summary <- qcSummary(rfastp_report)
df_summary
```

# Batch file processing

That's nice, but we rarely just have a single fastq file, and we'd like to look at them all at once. Luckily, we can do that with rfastp.

First, we need to get the locations of all of the files we downloaded earlier

```{r}
# adjust to the path where you assigned in DATA_DIR above (replace /Users/$USER by default is same as ~)
fq_file_dir <- "~/Desktop/Genomic_Data_Analysis"
# crate a list of all of the files
fastq.files <- list.files(path = fq_file_dir, # where to look
                          pattern = "rep[0-9].fastq.gz$", # the pattern of file name to find
                                  # Note, if you have other fastq files in the folder, they will also be included.
                          full.names = TRUE) # save the full path to the file

print(fastq.files)
```

Now we have all of the file paths

We can loop through all of the files to perform filtering and trimming. Note there are many arguments that can be modified. Use ?rfastp to learn more.

```{r}
# create a directory for the output to go into if not already present
output_dir <- paste0(fq_file_dir, "/filtered") 
if (!dir.exists(output_dir)) {dir.create(output_dir, recursive = TRUE)}

# run rfastp on all fastq files
for (i in 1:length(fastq.files)) {
  # file path to single end read
  read1 <- fastq.files[i]
  # assign output file (putting it inside of filtered folder)
  output_name <- paste0(dirname(fastq.files[i]),
                        "/filtered/",
                        "filtered_",
                        basename(fastq.files[i]))
  json_report <- rfastp(
    read1 = read1,
    outputFastq = paste0(str_split(output_name, fixed("."))[[1]][1],"_rfastp"),
    disableTrimPolyG = TRUE,
    # cutLowQualFront = TRUE,
    # cutFrontWindowSize = 3,
    # cutFrontMeanQual = 10,
    # cutLowQualTail = TRUE,
    cutTailWindowSize = 1,
    # cutTailMeanQual = 5,
    minReadLength = 15,
    # trimFrontRead1 = 10,
    # adapterSequenceRead1 = 'GTGTCAGTCACTTCCAGCGG'
  )
  
  # Print the output file link in the R Markdown document
  cat(paste0(
    "[Processing Complete - ",
    basename(output_name),
    "](",
    output_name,
    ")\n\n"
  ))
}

```

## QC and adapters

Another common tool for quality control is called FastQC, useable via command line or GUI, available at <https://www.bioinformatics.babraham.ac.uk/projects/fastqc/> or via pip or conda install in the command line.

To use this tool, let's get conda running on your computer. NOTE: Anaconda is already installed on computers in the computer lab. If you are using your own computer, you'll need to have conda installed ([link to learn more](https://docs.conda.io/en/main/miniconda.html))

First, we need to open a terminal window. We will copy code from the below code chunk into the terminal window.

```{bash initialize-conda, eval=FALSE}
. /opt/anaconda3/bin/activate && conda init
#. /opt/anaconda3/bin/activate && conda activate /opt/anaconda3;

# run this command in terminal to make sure conda is activated
which conda

# copy these 4 lines into terminal and run them
conda config --add channels defaults
conda config --append channels bioconda
conda config --append channels conda-forge
conda config --set channel_priority strict

# check channel order
conda config --show channels
```

```{bash create-condaEnvironment, engine.opts='-l'}
# create an enviornment for our QC packages
if conda info --envs | grep -q QC; then echo "enviornment 'QC' already exists"; else conda create -y -n QC fastqc multiqc; fi

# see available conda environments
conda env list

# activate our QC environment
conda activate QC

# make sure desired packages are working
which fastqc

which multiqc

```

# 

```{bash fastqc, engine.opts='-l'}
# Be sure to change this file path to the path you want your data to go
DATA_DIR="/Users/$USER/Desktop"

# Activate conda QC environment
conda activate QC

# show which version of fastqc is active
fastqc -v

# Function to check if a command is installed
command_exists() {
  command -v "$1" >/dev/null 2>&1
}

if command_exists fastqc; then
  # Run the command if fastqc is installed
  fastqc $DATA_DIR/*.fastq.gz -o $DATA_DIR/fastqc/raw
  fastqc $DATA_DIR/filtered/*.fastq.gz -o $DATA_DIR/fastqc/filtered
  if [ $? -ne 0 ]; then
    echo "FastQC execution failed. It didn't work."
  fi
else
  echo "FastQC is not installed."
fi
```

This link shows the fastqc output for subset_phenol_unstressed_rep4.fastq.gz

```{r include-fastQC, eval=FALSE}
browseURL("~/Desktop/fastqc/filtered/filtered_subset_phenol_unstressed_rep4_rfastp_R1_fastqc.html")
# htmltools::includeHTML("~/Desktop/filtered/fastqc/filtered_subset_phenol_unstressed_rep4_rfastp_R1_fastqc.html")
```

```{r fastqc-file, eval=FALSE}
browseURL("https://htmlpreview.github.io/?https://github.com/clstacy/GenomicDataAnalysis_Fa23/blob/main/outputs/fastqc/subset_phenol_unstressed_rep4_fastqc.html")
```

One of our favorite ways to analyze multiple samples simultaneously is [MultiQC](https://multiqc.info/) a software that combines fastQC reports

Here is the code to run it:

```{bash fastqc, engine.opts='-l'}
# Be sure to change this file path to the path you want to run multiqc
DATA_DIR="/Users/$USER/Desktop"

# activate QC environment
conda activate QC

# run multiqc on all of the fastqc outputs
multiqc $DATA_DIR/fastqc -o $DATA_DIR/multiqc -m fastqc -f
```

```{r multiqc-file, eval=FALSE}
browseURL("~/Desktop/mul")
browseURL("https://htmlpreview.github.io/?https://github.com/clstacy/GenomicDataAnalysis_Fa23/blob/main/outputs/fastqc/multiqc_report.html")
```

Be sure to knit this file into a pdf or html file once you're finished.

System information for reproducibility:

```{r}
pander::pander(sessionInfo())
```
